{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "This is a different approach to the assembly gap/break analysis that was not used\n",
    "in the Science 2021 paper (already includes T2T as reference assembly)\n",
    "- this notebook aggregates data on the basis of the contig to reference alignment BED files\n",
    "and produces filtered output for the corresponding eval notebook\n",
    "\"\"\"\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "path = '/home/local/work/data/hgsvc/contig_aln_bed/unfiltered'\n",
    "path = '/home/local/work/data/hgsvc/contig_aln_bed_t2t'\n",
    "out_lenient = '/home/local/work/data/hgsvc/contig_aln_bed/unfiltered/grt00'\n",
    "out_lenient = '/home/local/work/data/hgsvc/contig_aln_bed_t2t/grt00'\n",
    "out_stringent = '/home/local/work/data/hgsvc/contig_aln_bed/unfiltered/grt20'\n",
    "out_stringent = '/home/local/work/data/hgsvc/contig_aln_bed_t2t/grt20'\n",
    "os.makedirs(out_lenient, exist_ok=True)\n",
    "os.makedirs(out_stringent, exist_ok=True)\n",
    "\n",
    "names = ['chrom', 'start', 'end', 'name', 'mapq', 'strand']\n",
    "chrom_match = 'chr[0-9XY]+$'\n",
    "\n",
    "for bedfile in os.listdir(path):\n",
    "    if not bedfile.endswith('.bed'):\n",
    "        continue\n",
    "    file_path = os.path.join(path, bedfile)\n",
    "    if 'h1-un' in bedfile:\n",
    "        hap = 'H1'\n",
    "    else:\n",
    "        hap = 'H2'\n",
    "    sample, _, platform = bedfile.split('_map-to_')[0].rsplit('_', 1)[0].split('_')\n",
    "    if 'clr' in platform:\n",
    "        platform = 'CLR'\n",
    "    else:\n",
    "        platform = 'HiFi'\n",
    "    bed = pd.read_csv(file_path, sep='\\t', header=None, names=names)\n",
    "    bed = bed.loc[bed['chrom'].str.match(chrom_match), :].copy()\n",
    "    bed['name'] = sample + '_' + platform + '_' + hap + '_' + bed['name']\n",
    "    \n",
    "    bed_lenient = bed.loc[bed['mapq'] > 0, :]\n",
    "    file_lenient = bedfile.replace('.bed', '.mq-grt00.bed')\n",
    "    path_lenient = os.path.join(out_lenient, file_lenient)\n",
    "    bed_lenient.to_csv(path_lenient, sep='\\t', header=False, index=False)\n",
    "    \n",
    "    bed_stringent = bed.loc[bed['mapq'] > 20, :]\n",
    "    file_stringent = bedfile.replace('.bed', '.mq-grt20.bed')\n",
    "    path_stringent = os.path.join(out_stringent, file_stringent)\n",
    "    bed_stringent.to_csv(path_stringent, sep='\\t', header=False, index=False)\n",
    "\n",
    "\n",
    "def load_all_complements(folder):\n",
    "    \n",
    "    #children = ['HG00733', 'NA19240', 'HG00514']\n",
    "    loaded = []\n",
    "    bed_files = [os.path.join(folder, b) for b in os.listdir(folder) if b.endswith('complement.bed')]\n",
    "    for b in bed_files:\n",
    "        sample = os.path.basename(b).split('_')[0]\n",
    "        #if sample in children:\n",
    "        #    continue\n",
    "        if 'h1-un' in b:\n",
    "            hap = 'H1'\n",
    "        else:\n",
    "            hap = 'H2'\n",
    "        if 'pbsq2-ccs' in b:\n",
    "            platform = 'HiFi'\n",
    "        else:\n",
    "            platform = 'CLR'\n",
    "        name = '_'.join([sample, platform, hap])\n",
    "        df = pd.read_csv(b, sep='\\t', header=None, names=['chrom', 'start', 'end'])\n",
    "        df['name'] = name\n",
    "        loaded.append(df)\n",
    "    loaded = pd.concat(loaded, axis=0, ignore_index=False)\n",
    "    loaded.sort_values(['chrom', 'start', 'end'], inplace=True)\n",
    "    loaded.reset_index(drop=True, inplace=True)\n",
    "    return loaded\n",
    "    \n",
    "mq00_files = load_all_complements(out_lenient)\n",
    "mq00_merged_out = os.path.join(out_lenient, 'all_grt00_concat.tsv')\n",
    "mq00_files.to_csv(mq00_merged_out, sep='\\t', header=False, index=False)\n",
    "\n",
    "mq20_files = load_all_complements(out_stringent)\n",
    "mq20_merged_out = os.path.join(out_stringent, 'all_grt20_concat.tsv')\n",
    "mq20_files.to_csv(mq20_merged_out, sep='\\t', header=False, index=False)\n",
    "\n",
    "def unflatten_merged_table(file_path, name_prefix):\n",
    "    \n",
    "    records = []\n",
    "    current_chrom = None\n",
    "    region_counter = 0\n",
    "    with open(file_path, 'r') as table:\n",
    "        for line in table:\n",
    "            clr_count = 0\n",
    "            hifi_count = 0\n",
    "            chrom, start, end, haplotypes = line.strip().split('\\t')\n",
    "            if current_chrom != chrom:\n",
    "                region_counter = 0\n",
    "                current_chrom = chrom\n",
    "            region_counter += 1\n",
    "            entry = {\n",
    "                'chrom': chrom,\n",
    "                'start': int(start),\n",
    "                'end': int(end),\n",
    "                'name': '_'.join([name_prefix, str(region_counter), start]),\n",
    "                'length': int(end) - int(start)\n",
    "            }\n",
    "            for h in haplotypes.split(','):\n",
    "                if 'CLR' in h:\n",
    "                    clr_count += 1\n",
    "                else:\n",
    "                    hifi_count += 1\n",
    "                entry[h] = 1\n",
    "            entry['COUNT_CLR'] = clr_count\n",
    "            entry['COUNT_HIFI'] = hifi_count\n",
    "            records.append(entry)\n",
    "    df = pd.DataFrame.from_records(records)\n",
    "    df.fillna(0, inplace=True)\n",
    "    column_typer = dict((c, int) for c in df.columns if c not in ['chrom', 'name'])\n",
    "    column_typer['chrom'] = str\n",
    "    df = df.astype(column_typer)\n",
    "    col_sort = ['chrom', 'start', 'end', 'name', 'length', 'COUNT_CLR', 'COUNT_HIFI']\n",
    "    sorted_samples = sorted([c for c in df.columns if c not in col_sort])\n",
    "    col_sort.extend(sorted_samples)\n",
    "    df = df[col_sort]\n",
    "    return df\n",
    "\n",
    "file_paths = [\n",
    "    '/home/local/work/data/hgsvc/contig_aln_bed/unfiltered/grt20/all_grt20_merged.tsv',\n",
    "    '/home/local/work/data/hgsvc/contig_aln_bed/unfiltered/grt20/all_grt20_no-child_merged.tsv',\n",
    "    '/home/local/work/data/hgsvc/contig_aln_bed/unfiltered/grt00/all_grt00_merged.tsv',\n",
    "    '/home/local/work/data/hgsvc/contig_aln_bed/unfiltered/grt00/all_grt00_no-child_merged.tsv',\n",
    "    '/home/local/work/data/hgsvc/contig_aln_bed_t2t/grt00/all_grt00_merged.tsv',\n",
    "    '/home/local/work/data/hgsvc/contig_aln_bed_t2t/grt00/all_grt00_no-child_merged.tsv',\n",
    "    '/home/local/work/data/hgsvc/contig_aln_bed_t2t/grt20/all_grt20_merged.tsv',\n",
    "    '/home/local/work/data/hgsvc/contig_aln_bed_t2t/grt20/all_grt20_no-child_merged.tsv',\n",
    "]\n",
    "\n",
    "name_prefixes = [\n",
    "    'GRCh38_MQ20_H70',\n",
    "    'GRCh38_MQ20_H64',\n",
    "    'GRCh38_MQ00_H70',\n",
    "    'GRCh38_MQ00_H64',\n",
    "    'T2T_MQ00_H70',\n",
    "    'T2T_MQ00_H64',\n",
    "    'T2T_MQ20_H70',\n",
    "    'T2T_MQ20_H64',\n",
    "]\n",
    "\n",
    "for fp, np in zip(file_paths, name_prefixes):\n",
    "    df = unflatten_merged_table(fp, np)\n",
    "    out_file = os.path.join(os.path.dirname(fp), 'Gaps_' + np + '.tsv')\n",
    "    with open(out_file, 'w') as dump:\n",
    "        _ = dump.write('#')\n",
    "        df.to_csv(dump, sep='\\t', header=True, index=False)\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
