{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cluster16_contig_129\n",
      "(3, 13)\n",
      "cluster2_contig_16\n",
      "(15, 19)\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import io\n",
    "import re\n",
    "import collections as col\n",
    "\n",
    "\"\"\"\n",
    "What does this do?\n",
    "Cut out sequence for unknown/black segment from phased assembly contigs\n",
    "to align against GRCh38, and to perform pairwise sequence alignment checking\n",
    "for broad homology between unknown/black segments coming from different assemblies.\n",
    "Currently, not used or mentioned in the manuscript.\n",
    "\"\"\"\n",
    "\n",
    "segments = \"\"\"\n",
    "bng_row_index   sample population platform assembly_type haplotype  \\\n",
    "110             45  HG02011        AFR      CLR        phased        H1   \n",
    "189             65  NA19238        AFR      CLR        phased        H2   \n",
    "\n",
    "    bng_contig_id  component_start  component_end orientation  \\\n",
    "110        h1-649               -1         160528      direct   \n",
    "189       h2-1014             7554         191560      direct   \n",
    "\n",
    "     component_length             contig_id variant_type  \\\n",
    "110           2389451  cluster16_contig_129   no_variant   \n",
    "189           2416853    cluster2_contig_16   no_variant   \n",
    "\n",
    "    variant_segment_chain  variant_start  variant_end segment_color  \\\n",
    "110      no_segment_chain             -1           -1         black   \n",
    "189      no_segment_chain             -1           -1         black   \n",
    "\n",
    "     segment_start  segment_end  segment_support  \n",
    "110          53703        99375                2  \n",
    "189          41551        88156                2\n",
    "\"\"\"\n",
    "\n",
    "path = '/home/local/work/data/hgsvc/roi/dotplots_chr3/chr3_ctg_fasta'\n",
    "fasta_files = [\n",
    "    'HG02011_hgsvc_pbsq2-clr_1000-flye.h1-un.arrow-p1.ctg3q29.fasta',\n",
    "    'NA19238_hgsvc_pbsq2-clr_1000-flye.h2-un.arrow-p1.ctg3q29.fasta'\n",
    "]\n",
    "\n",
    "def count_motifs(sequence):\n",
    "    \n",
    "    forward = re.findall('CTTAAG', sequence.upper())\n",
    "    reverse = re.findall('GAATTC', sequence.upper())\n",
    "    return len(forward), len(reverse)\n",
    "\n",
    "\n",
    "seq_buffer = io.StringIO()\n",
    "contig_buffer = ''\n",
    "buffered_contig = ''\n",
    "record = False\n",
    "for f in fasta_files:\n",
    "    fpath = os.path.join(path, f)\n",
    "    with open(fpath, 'r') as fasta:\n",
    "        for line in fasta:\n",
    "            if line.startswith('>'):\n",
    "                contig = line.strip('>').strip()\n",
    "                if contig in ['cluster16_contig_129', 'cluster2_contig_16']:\n",
    "                    record = True\n",
    "                if contig_buffer:\n",
    "                    if buffered_contig == 'cluster16_contig_129':\n",
    "                        contig_buffer = contig_buffer[53703:99375]\n",
    "                        print(buffered_contig)\n",
    "                        print(count_motifs(contig_buffer))\n",
    "                        seq_buffer.write('>{}_black'.format(buffered_contig) + '\\n')\n",
    "                        seq_buffer.write(contig_buffer + '\\n')\n",
    "                    elif buffered_contig == 'cluster2_contig_16':\n",
    "                        contig_buffer = contig_buffer[41551:88156]\n",
    "                        print(buffered_contig)\n",
    "                        print(count_motifs(contig_buffer))\n",
    "                        seq_buffer.write('>{}_black'.format(buffered_contig) + '\\n')\n",
    "                        seq_buffer.write(contig_buffer + '\\n')\n",
    "                    else:\n",
    "                        pass\n",
    "                contig_buffer = ''\n",
    "                if record:\n",
    "                    buffered_contig = contig\n",
    "                continue\n",
    "            if record:\n",
    "                contig_buffer += line.strip()\n",
    "if contig_buffer:\n",
    "    if buffered_contig == 'cluster16_contig_129':\n",
    "        contig_buffer = contig_buffer[53703:99375]\n",
    "        seq_buffer.write('>{}_black'.format(buffered_contig) + '\\n')\n",
    "        seq_buffer.write(contig_buffer + '\\n')\n",
    "    elif buffered_contig == 'cluster2_contig_16':\n",
    "        contig_buffer = contig_buffer[41551:88156]\n",
    "        print(buffered_contig)\n",
    "        print(count_motifs(contig_buffer))\n",
    "        seq_buffer.write('>{}_black'.format(buffered_contig) + '\\n')\n",
    "        seq_buffer.write(contig_buffer + '\\n')\n",
    "\n",
    "with open(os.path.join(path, 'HG02011-H1-CLR_NA19238-H2-CLR_black.fasta'), 'w') as fasta:\n",
    "    _ = fasta.write(seq_buffer.getvalue())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
