{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/local/software/conda/envs/nbsrv/lib/python3.7/site-packages/ipykernel_launcher.py:203: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "ename": "RuntimeError",
     "evalue": "No active exception to reraise",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-16-c96e76f4b5f9>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m    211\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    212\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 213\u001b[0;31m \u001b[0;32mraise\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    214\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    215\u001b[0m raise RuntimeError('Code beyond this point expects data on the '\n",
      "\u001b[0;31mRuntimeError\u001b[0m: No active exception to reraise"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import collections as col\n",
    "import importlib\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as patches\n",
    "import matplotlib.lines as lines\n",
    "\n",
    "plot_aux_module = '/home/local/work/code/github/project-diploid-assembly/notebooks/aux_mods/plot_aux.py'\n",
    "plot_aux_spec = importlib.util.spec_from_file_location(\"plot_aux\", plot_aux_module)\n",
    "plot_aux = importlib.util.module_from_spec(plot_aux_spec)\n",
    "plot_aux_spec.loader.exec_module(plot_aux)\n",
    "\n",
    "samples = plot_aux.load_sample_table()\n",
    "hexcodes, rgbcodes, popmap = plot_aux.load_population_annotation()\n",
    "\n",
    "save_plot = False\n",
    "print_stats = False\n",
    "dump_minimal_summary = False\n",
    "dump_tableS9_columns = True\n",
    "make_test = False\n",
    "dump_length_stats = False\n",
    "rel_hap_n50 = False\n",
    "\n",
    "prop = plot_aux.get_plot_property\n",
    "\n",
    "pipeline_version = 'v12'\n",
    "\n",
    "# color_version\n",
    "# - full\n",
    "# - super_pop\n",
    "# - red_black\n",
    "color_version = 'red_black'\n",
    "\n",
    "if color_version == 'full' or color_version == 'super_pop':\n",
    "    rgbcodes['HiFi'] = tuple(map(lambda x: round(x/255, 2), (35, 35, 35)))\n",
    "    rgbcodes['CLR'] = tuple(map(lambda x: round(x/255, 2), (35, 35, 35)))\n",
    "elif color_version == 'red_black':\n",
    "    rgbcodes['HiFi'] = plot_aux.get_platform_color('HiFi')\n",
    "    rgbcodes['CLR'] = plot_aux.get_platform_color('CLR')\n",
    "else:\n",
    "    raise\n",
    "\n",
    "stats_path = '/home/local/work/data/hgsvc/fig1_panels/busco_stats'\n",
    "out_path = os.path.join(stats_path, 'quast_busco_table.tsv')\n",
    "cache_file = os.path.join(stats_path, 'cache_{}.h5'.format(pipeline_version))\n",
    "\n",
    "# value reported by QUAST (in \"quast.log\")\n",
    "# features of type \"gene\" in GENCODE v31 basic\n",
    "total_gencode_genes = 60603\n",
    "\n",
    "keep_infos = {\n",
    "    4: 'num_contigs_all',\n",
    "    10: 'total_length_bp',\n",
    "    16: 'num_contigs_geq_3kb',\n",
    "    17: 'largest_contig_bp',\n",
    "    18: 'total_length_geq3kb_bp',\n",
    "    20: 'GC_pct',\n",
    "    22: 'N50',\n",
    "    23: 'NG50',\n",
    "    24: 'N75',\n",
    "    25: 'NG75',\n",
    "    30: 'num_misassemblies',\n",
    "    40: 'ref_genome_fraction_pct',\n",
    "    41: 'dup_ratio',\n",
    "    42: 'Nper100kbp_ratio',\n",
    "    45: 'gencode_genes',\n",
    "    46: 'busco_comp_pct',\n",
    "    47: 'busco_part_pct',\n",
    "    48: 'largest_alignment_bp',\n",
    "    49: 'total_aligned_length_bp'\n",
    "}\n",
    "\n",
    "def load_assembly_stats(file_path):\n",
    "    assm_stats = dict()\n",
    "    with open(file_path, 'r') as table:\n",
    "        for ln, line in enumerate(table, start=1):\n",
    "            if ln not in keep_infos:\n",
    "                continue\n",
    "            key = keep_infos[ln]\n",
    "            parts = line.strip().split()\n",
    "            if key == 'gencode_genes':\n",
    "                key_complete = key + '_comp_num'\n",
    "                key_partial = key + '_part_num'\n",
    "                complete_genes = int(parts[-4])\n",
    "                partial_genes = int(parts[-2])\n",
    "                assm_stats[key_complete] = complete_genes\n",
    "                assm_stats[key_partial] = partial_genes\n",
    "                cmp_genes_pct = round(complete_genes / total_gencode_genes * 100, 2)\n",
    "                part_genes_pct = round(partial_genes / total_gencode_genes * 100, 2)\n",
    "                assm_stats[key_complete.replace('_num', '_pct')] = cmp_genes_pct\n",
    "                assm_stats[key_partial.replace('_num', '_pct')] = part_genes_pct\n",
    "                continue\n",
    "            to_num = int\n",
    "            if any(key.endswith(x) for x in ['pct', 'ratio']):\n",
    "                to_num = float\n",
    "            try:\n",
    "                assm_stats[key] = to_num(parts[-1])\n",
    "            except ValueError:\n",
    "                raise ValueError('Cannot convert number: {} / {} / {}'.format(ln, line.strip(), file_path))\n",
    "    return assm_stats\n",
    "\n",
    "\n",
    "def extract_assembly_type(file_name):\n",
    "    assmt = None\n",
    "    if '_nhr-' in file_name:\n",
    "        assmt = 'NHR', 0\n",
    "    elif 'h1-un' in file_name:\n",
    "        assmt = 'HAP', 10\n",
    "    elif 'h2-un' in file_name:\n",
    "        assmt = 'HAP', 20\n",
    "    else:\n",
    "        raise ValueError('ASSMT: {}'.format(file_name))\n",
    "    if assmt is None:\n",
    "        raise ValueError('ASSMT: {}'.format(file_name))\n",
    "    return assmt\n",
    "\n",
    "\n",
    "if not os.path.isfile(cache_file):\n",
    "    stat_files = plot_aux.load_plot_data_files(\n",
    "        stats_path,\n",
    "        '.txt',\n",
    "        pipeline_version)\n",
    "    plot_data = [load_assembly_stats(fp) for fp in stat_files]\n",
    "\n",
    "    sample_platform = [plot_aux.extract_sample_platform(os.path.basename(f)) for f in stat_files]\n",
    "    assembly_types = [extract_assembly_type(os.path.basename(f)) for f in stat_files]\n",
    "    row_index = []\n",
    "    for (sample, platform), (assmt, hap) in zip(sample_platform, assembly_types):\n",
    "        super_pop = samples[sample]['super_population']\n",
    "        pop = samples[sample]['population']\n",
    "        row_index.append((sample, super_pop, pop, platform, assmt, hap))\n",
    "\n",
    "    df = pd.DataFrame(\n",
    "        plot_data,\n",
    "        columns=sorted(plot_data[0].keys()),\n",
    "        index=row_index\n",
    "    )\n",
    "\n",
    "    df.index = pd.MultiIndex.from_tuples(\n",
    "        df.index.values,\n",
    "        names=['sample', 'super_pop', 'pop', 'platform', 'assembly', 'hap']\n",
    "    )\n",
    "    df.to_hdf(cache_file, key='cache', mode='w', format='fixed')\n",
    "    \n",
    "df = pd.read_hdf(cache_file, 'cache')\n",
    "df.sort_index(axis=0, inplace=True, level=['sample', 'super_pop', 'pop', 'platform', 'assembly', 'hap'])        \n",
    "\n",
    "errors, missing = plot_aux.check_cache_consistency(df)\n",
    "if errors:\n",
    "    raise ValueErrors('Skip sample(s) in cache: {}'.format(errors))\n",
    "if missing:\n",
    "    print(sorted(missing))\n",
    "    print('---------------')\n",
    "\n",
    "if print_stats:\n",
    "    print(df['gencode_genes_comp_pct'].describe())\n",
    "    print(df['gencode_genes_part_pct'].describe())\n",
    "    print(df['busco_comp_pct'].describe())\n",
    "    print(df['busco_part_pct'].describe())\n",
    "    print(df['Nper100kbp_ratio'].describe())\n",
    "\n",
    "columns = [\n",
    "    'busco_comp_pct',\n",
    "    'busco_part_pct',\n",
    "    'gencode_genes_comp_pct',\n",
    "    'gencode_genes_part_pct',\n",
    "    'Nper100kbp_ratio',\n",
    "    'GC_pct',\n",
    "]\n",
    "summary_header = [\n",
    "    'Technology',\n",
    "    'BUSCO (complete, %)',\n",
    "    'BUSCO (partial, %)',\n",
    "    'GENCODEv31 genes (complete, %)',\n",
    "    'GENCODEv31 genes (partial, %)',\n",
    "    'N per 100 kbp',\n",
    "    'GC content (%)',\n",
    "]\n",
    "\n",
    "    \n",
    "if dump_minimal_summary:\n",
    "\n",
    "    mini_summary = []\n",
    "    mini_summary.append('\\t'.join(summary_header))\n",
    "    for selector in ['CLR', 'HiFi']:\n",
    "        sub = df.xs(selector, level='platform', axis=0)\n",
    "        sub = sub[columns]\n",
    "        stat_line = [selector]\n",
    "        for avg, med in zip(sub.mean(axis=0), sub.median(axis=0)):\n",
    "            avg = round(avg, 1)\n",
    "            med = round(med, 1)\n",
    "            stat_line.append('{} / {}'.format(avg, med))\n",
    "        mini_summary.append('\\t'.join(stat_line))\n",
    "    with open('/home/ebertp/mini_summary.tsv', 'w') as dump:\n",
    "        dump.write('\\n'.join(mini_summary) + '\\n')\n",
    "        \n",
    "if dump_tableS9_columns:\n",
    "    sub = df[columns]\n",
    "    sub.sort_index(level=['platform', 'sample', 'hap'], ascending=[True, True, True], inplace=True)\n",
    "    out_tsv = '/home/ebertp/tableS9_ext.tsv'\n",
    "    sub.to_csv(\n",
    "        out_tsv,\n",
    "        sep='\\t',\n",
    "        index=True,\n",
    "        header=True\n",
    "    )\n",
    "    \n",
    "\n",
    "raise\n",
    "\n",
    "raise RuntimeError('Code beyond this point expects data on the '\n",
    "                   'non-haplotype resolved (NHR) assemblies, which is '\n",
    "                   'not available for the QUAST/BUSCO evaluation runs.')\n",
    "\n",
    "def plot_assembly_contiguity(sample_stats):\n",
    "        \n",
    "    nhr_assm = sample_stats.xs('NHR', level='assembly', drop_level=False)\n",
    "    \n",
    "    x_vals = col.defaultdict(list)\n",
    "    y_vals = col.defaultdict(list)\n",
    "    colors = col.defaultdict(list)\n",
    "    \n",
    "    for idx, row in nhr_assm.iterrows():\n",
    "        platform = idx[3]\n",
    "        population = idx[2]\n",
    "        super_pop = idx[1]\n",
    "        \n",
    "        sqa_n50 = round(row['N50'] / 1e6, 2)\n",
    "        x_vals[platform].append(sqa_n50)\n",
    "        \n",
    "        hap1_index = (idx[0], idx[1], idx[2], idx[3], 'HAP', 10)\n",
    "        hap1_n50 = round(sample_stats.loc[hap1_index, 'N50'] / 1e6, 2)\n",
    "        if rel_hap_n50:\n",
    "            hap1_n50 = round(hap1_n50 / sqa_n50, 2)\n",
    "\n",
    "        y_vals[(platform, 'HAP1')].append(hap1_n50)\n",
    "        \n",
    "        hap2_index = (idx[0], idx[1], idx[2], idx[3], 'HAP', 20)\n",
    "        hap2_n50 = round(sample_stats.loc[hap2_index, 'N50'] / 1e6, 2)\n",
    "        if rel_hap_n50:\n",
    "            hap2_n50 = round(hap2_n50 / sqa_n50, 2)\n",
    "\n",
    "        y_vals[(platform, 'HAP2')].append(hap2_n50)\n",
    "        \n",
    "        if color_version == 'full':\n",
    "            colors[platform].append(rgbcodes[population])\n",
    "        elif color_version == 'super_pop':\n",
    "            colors[platform].append(rgbcodes[super_pop])\n",
    "        elif color_version == 'red_black':\n",
    "            colors[platform].append(rgbcodes[platform])\n",
    "        else:\n",
    "            raise\n",
    "        \n",
    "    \n",
    "    fig, axis = plt.subplots(figsize=(8,8))\n",
    "    \n",
    "    for platform in plot_aux.get_sequencing_platforms():\n",
    "        for hap in ['HAP1', 'HAP2']:\n",
    "            nhr_n50 = x_vals[platform]\n",
    "            hap_n50 = y_vals[(platform, hap)]\n",
    "            marker_colors = colors[platform]\n",
    "\n",
    "            axis.scatter(\n",
    "                nhr_n50,\n",
    "                hap_n50,\n",
    "                s=prop('plot_marker_size'),\n",
    "                c=colors[platform],\n",
    "                marker=prop('{}_marker'.format(platform))\n",
    "            )\n",
    "    \n",
    "    axis.set_xlabel(\n",
    "        'Squashed assembly contig N50 (Mbp)',\n",
    "        fontsize=prop('fontsize_legend')\n",
    "    )\n",
    "    if rel_hap_n50:\n",
    "        axis.set_ylabel(\n",
    "            'Haploid assembly contig N50 (relative units)',\n",
    "            fontsize=prop('fontsize_legend')\n",
    "        )\n",
    "    else:\n",
    "        axis.set_ylabel(\n",
    "            'Haploid assembly contig N50 (Mbp)',\n",
    "            fontsize=prop('fontsize_legend')\n",
    "        )\n",
    "    axis.spines['top'].set_visible(False)\n",
    "    axis.spines['right'].set_visible(False)\n",
    "    \n",
    "    axis.tick_params(\n",
    "        axis='both',\n",
    "        which='major',\n",
    "        labelsize=prop('fontsize_axis_ticks')\n",
    "    )\n",
    "    \n",
    "    # build custom legend\n",
    "    custom_lines = [\n",
    "        lines.Line2D(\n",
    "            [0], [0],\n",
    "            color=rgbcodes['CLR'],\n",
    "            markersize=prop('legend_marker_size'),\n",
    "            marker=prop('CLR_marker'),\n",
    "            ls='None',\n",
    "            label='CLR'\n",
    "        ),\n",
    "        lines.Line2D(\n",
    "            [0], [0],\n",
    "            color=rgbcodes['HiFi'],\n",
    "            markersize=prop('legend_marker_size'),\n",
    "            marker=prop('HiFi_marker'),\n",
    "            ls='None',\n",
    "            label='HiFi'\n",
    "        )   \n",
    "    ]\n",
    "    \n",
    "    axis.legend(\n",
    "        handles=custom_lines,\n",
    "        loc='upper left',\n",
    "        prop={'size': prop('fontsize_legend')}\n",
    "    )\n",
    "    \n",
    "    axis.plot(\n",
    "        [0, 1],\n",
    "        [0, 1],\n",
    "        color='darkgrey',\n",
    "        linestyle='dotted',\n",
    "        lw='2',\n",
    "        transform=axis.transAxes\n",
    "    )\n",
    "    axis.set_xlim(10, 40)\n",
    "    axis.set_ylim(10, 40)\n",
    "\n",
    "    if missing:\n",
    "        plot_aux.add_incomplete_stamp(axis, 0.5, 0.85)\n",
    "    \n",
    "    extra_artists = []\n",
    "    \n",
    "    return fig, extra_artists\n",
    "        \n",
    "fig, exart = plot_assembly_contiguity(df)\n",
    "\n",
    "\n",
    "if save_plot:\n",
    "    if color_version == 'red_black':\n",
    "        out_path += '_rb'\n",
    "    if color_version == 'super_pop':\n",
    "        out_path += '_spop'\n",
    "    if rel_hap_n50:\n",
    "        output += '_rel'\n",
    "    fig.savefig(out_path + '.png', dpi=600, bbox_inches='tight', extra_artists=exart)\n",
    "    fig.savefig(out_path + '_lowres.png', dpi=150, bbox_inches='tight', extra_artists=exart)\n",
    "    fig.savefig(out_path + '.svg', bbox_inches='tight', extra_artists=exart)\n",
    "\n",
    "if dump_length_stats:\n",
    "    length_stats = os.path.join(stats_path, 'assembly_lengths.tsv')\n",
    "    df.sort_index(axis=0, level=['platform', 'sample', 'super_pop', 'pop', 'hap'], inplace=True)\n",
    "    df.to_csv(\n",
    "        length_stats,\n",
    "        sep='\\t',\n",
    "        header=True,\n",
    "        index=True,\n",
    "        columns=['total_length_bp'],\n",
    "        mode='w'\n",
    "    )\n",
    "    \n",
    "if print_stats:\n",
    "    print('=== STATS SUMMARY ===')\n",
    "    nhr_all = df.xs('NHR', level='assembly', drop_level=False)\n",
    "    print('NHR ALL stats')\n",
    "    print(nhr_all['N50'].describe())\n",
    "    print('==========')\n",
    "    nhr_clr = df.xs(['CLR', 'NHR'], level=['platform', 'assembly'], drop_level=False)\n",
    "    print('NHR CLR stats')\n",
    "    print(nhr_clr['N50'].describe())\n",
    "    print('% > 35: ', round((nhr_clr['N50'] > 35e6).sum() / nhr_clr.shape[0] * 100, 1))\n",
    "    print('% > 30: ', round((nhr_clr['N50'] > 30e6).sum() / nhr_clr.shape[0] * 100, 1))\n",
    "    print('==========')\n",
    "    nhr_hifi = df.xs(['HiFi', 'NHR'], level=['platform', 'assembly'], drop_level=False)\n",
    "    print('NHR HiFi stats')\n",
    "    print(nhr_hifi['N50'].describe())\n",
    "    print('% > 35: ', round((nhr_hifi['N50'] > 35e6).sum() / nhr_hifi.shape[0] * 100, 1))\n",
    "    print('% > 30: ', round((nhr_hifi['N50'] > 30e6).sum() / nhr_hifi.shape[0] * 100, 1))\n",
    "    print('==========')\n",
    "\n",
    "    print('==========\\n=========')\n",
    "    hap_all = df.xs('HAP', level='assembly', drop_level=False)\n",
    "    print('HAP ALL stats')\n",
    "    print(hap_all['N50'].describe())\n",
    "    print('==========')\n",
    "    hap_clr = df.xs(['CLR', 'HAP'], level=['platform', 'assembly'], drop_level=False)\n",
    "    print('================')\n",
    "    print('HAP CLR stats')\n",
    "    print(hap_clr['N50'].describe())\n",
    "    print('% > 35: ', (hap_clr['N50'] > 35e6).sum() / hap_clr.shape[0] * 100)\n",
    "    print('% > 33: ', (hap_clr['N50'] > 33e6).sum() / hap_clr.shape[0] * 100)\n",
    "    print('% > 32: ', (hap_clr['N50'] > 32e6).sum() / hap_clr.shape[0] * 100)\n",
    "    print('% > 30: ', (hap_clr['N50'] > 30e6).sum() / hap_clr.shape[0] * 100)\n",
    "    print('# above 35 Mbp: ', (hap_clr['N50'] > 35e6).sum())\n",
    "    print('# above 33 Mbp: ', (hap_clr['N50'] > 33e6).sum())\n",
    "    print('# above 32 Mbp: ', (hap_clr['N50'] > 32e6).sum())\n",
    "    print('# above 30 Mbp: ', (hap_clr['N50'] > 30e6).sum())\n",
    "    print('==========')\n",
    "    hap_hifi = df.xs(['HiFi', 'HAP'], level=['platform', 'assembly'], drop_level=False)\n",
    "    print('================')\n",
    "    print('HAP HiFi stats')\n",
    "    print(hap_hifi['N50'].describe())\n",
    "    print('% > 35: ', (hap_hifi['N50'] > 35e6).sum() / hap_hifi.shape[0] * 100)\n",
    "    print('% > 33: ', (hap_hifi['N50'] > 33e6).sum() / hap_hifi.shape[0] * 100)\n",
    "    print('% > 32: ', (hap_hifi['N50'] > 32e6).sum() / hap_hifi.shape[0] * 100)\n",
    "    print('% > 30: ', (hap_hifi['N50'] > 30e6).sum() / hap_hifi.shape[0] * 100)\n",
    "    print('==========')\n",
    "    print('HAP HiFi non-AFR')\n",
    "    non_afr_index = hap_hifi.index.get_level_values('super_pop').isin(['AMR', 'EAS', 'EUR', 'SAS'])\n",
    "    hap_hifi_non_afr = hap_hifi.loc[non_afr_index, :].copy()\n",
    "    print(hap_hifi_non_afr['N50'].describe())\n",
    "    \n",
    "if make_test:\n",
    "    from scipy.stats import ttest_ind as ttest\n",
    "    from scipy.stats import bartlett\n",
    "    \n",
    "    hifi_hap = df.xs(['HAP', 'HiFi'], level=['assembly', 'platform'])\n",
    "    print(hifi_hap.shape)\n",
    "    clr_hap = df.xs(['HAP', 'CLR'], level=['assembly', 'platform'])\n",
    "    print(clr_hap.shape)\n",
    "    \n",
    "    hifi_n50 = hifi_hap['N50']\n",
    "    clr_n50 = clr_hap['N50']\n",
    "    \n",
    "    print(bartlett(hifi_n50, clr_n50))\n",
    "    \n",
    "    print(hifi_n50.var())\n",
    "    print(clr_n50.var())\n",
    "    \n",
    "    t_statistic, t_pv = ttest(\n",
    "        hifi_n50,\n",
    "        clr_n50,\n",
    "        equal_var=False\n",
    "    )\n",
    "    print('N HiFi: ', hifi_hap.shape[0])\n",
    "    print('N CLR: ', clr_hap.shape[0])\n",
    "    print(round(t_statistic, 3))\n",
    "    print(t_pv)\n",
    "    print(t_pv < 10e-9)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
