{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python\n",
    "\n",
    "\"\"\"\n",
    "Demonstrates plotting chromosome ideograms and genes (or any features, really)\n",
    "using matplotlib.\n",
    "1) Assumes a file from UCSC's Table Browser from the \"cytoBandIdeo\" table,\n",
    "saved as \"ideogram.txt\". Lines look like this::\n",
    "    #chrom  chromStart  chromEnd  name    gieStain\n",
    "    chr1    0           2300000   p36.33  gneg\n",
    "    chr1    2300000     5300000   p36.32  gpos25\n",
    "    chr1    5300000     7100000   p36.31  gneg\n",
    "2) Assumes another file, \"ucsc_genes.txt\", which is a BED format file\n",
    "   downloaded from UCSC's Table Browser. This script will work with any\n",
    "   BED-format file.\n",
    "\"\"\"\n",
    "import os as os\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.collections import BrokenBarHCollection\n",
    "import pandas\n",
    "import re as re\n",
    "\n",
    "__author__ = 'Ryan Dale'\n",
    "__source__ = 'https://www.biostars.org/p/147364/#147637'\n",
    "\n",
    "# Here's the function that we'll call for each dataframe (once for chromosome\n",
    "# ideograms, once for genes).  The rest of this script will be prepping data\n",
    "# for input to this function\n",
    "#\n",
    "def chromosome_collections(df, y_positions, height,  **kwargs):\n",
    "    \"\"\"\n",
    "    Yields BrokenBarHCollection of features that can be added to an Axes\n",
    "    object.\n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pandas.DataFrame\n",
    "        Must at least have columns ['chrom', 'start', 'end', 'color']. If no\n",
    "        column 'width', it will be calculated from start/end.\n",
    "    y_positions : dict\n",
    "        Keys are chromosomes, values are y-value at which to anchor the\n",
    "        BrokenBarHCollection\n",
    "    height : float\n",
    "        Height of each BrokenBarHCollection\n",
    "    Additional kwargs are passed to BrokenBarHCollection\n",
    "    \"\"\"\n",
    "    del_width = False\n",
    "    if 'width' not in df.columns:\n",
    "        del_width = True\n",
    "        df['width'] = df['end'] - df['start']\n",
    "    for chrom, group in df.groupby('chrom'):\n",
    "        yrange = (y_positions[chrom], height)\n",
    "        xranges = group[['start', 'width']].values\n",
    "        yield BrokenBarHCollection(\n",
    "            xranges, yrange, facecolors=group['colors'], **kwargs)\n",
    "    if del_width:\n",
    "        del df['width']\n",
    "\n",
    "\n",
    "# Height of each ideogram\n",
    "chrom_height = 1\n",
    "\n",
    "# Spacing between consecutive ideograms\n",
    "chrom_spacing = 1.5\n",
    "\n",
    "# Height of the gene track. Should be smaller than `chrom_spacing` in order to\n",
    "# fit correctly\n",
    "gene_height = 1\n",
    "\n",
    "# Padding between the top of a gene track and its corresponding ideogram\n",
    "gene_padding = 0.1\n",
    "\n",
    "# Width, height (in inches)\n",
    "figsize = (10, 12)\n",
    "\n",
    "# Decide which chromosomes to use\n",
    "chromosome_list = ['chr%s' % i for i in list(range(1, 23)) + ['M', 'X', 'Y']]\n",
    "chromosome_list = ['chr%s' % i for i in list(range(1, 23)) + ['X']]\n",
    "\n",
    "# Keep track of the y positions for ideograms and genes for each chromosome,\n",
    "# and the center of each ideogram (which is where we'll put the ytick labels)\n",
    "ybase = 0\n",
    "chrom_ybase = {}\n",
    "gene_ybase = {}\n",
    "chrom_centers = {}\n",
    "\n",
    "# Iterate in reverse so that items in the beginning of `chromosome_list` will\n",
    "# appear at the top of the plot\n",
    "for chrom in chromosome_list[::-1]:\n",
    "    chrom_ybase[chrom] = ybase\n",
    "    chrom_centers[chrom] = ybase + chrom_height / 2.\n",
    "    gene_ybase[chrom] = ybase - gene_height - gene_padding\n",
    "    ybase += chrom_height + chrom_spacing\n",
    "\n",
    "# Read in ideogram.txt, downloaded from UCSC Table Browser\n",
    "ideo = pandas.read_table(\n",
    "    '/home/pebert/Downloads/hg38_ideogram.bed',\n",
    "    skiprows=1,\n",
    "    names=['chrom', 'start', 'end', 'name', 'gieStain']\n",
    ")\n",
    "\n",
    "# Filter out chromosomes not in our list\n",
    "ideo = ideo[ideo.chrom.apply(lambda x: x in chromosome_list)]\n",
    "\n",
    "# Add a new column for width\n",
    "ideo['width'] = ideo.end - ideo.start\n",
    "\n",
    "# Colors for different chromosome stains\n",
    "color_lookup = {\n",
    "    'gneg': (1., 1., 1.),\n",
    "    'gpos25': (.6, .6, .6),\n",
    "    'gpos50': (.4, .4, .4),\n",
    "    'gpos75': (.2, .2, .2),\n",
    "    'gpos100': (0., 0., 0.),\n",
    "    'acen': (.8, .4, .4),\n",
    "    'gvar': (.8, .8, .8),\n",
    "    'stalk': (.9, .9, .9),\n",
    "}\n",
    "\n",
    "# Add a new column for colors\n",
    "ideo['colors'] = ideo['gieStain'].apply(lambda x: color_lookup[x])\n",
    "\n",
    "\n",
    "# # Same thing for genes\n",
    "# genes = pandas.read_table(\n",
    "#     'ucsc_genes.txt',\n",
    "#     names=['chrom', 'start', 'end', 'name'],\n",
    "#     usecols=range(4))\n",
    "# genes = genes[genes.chrom.apply(lambda x: x in chromosome_list)]\n",
    "# genes['width'] = genes.end - genes.start\n",
    "# genes['colors'] = '#2243a8'\n",
    "\n",
    "mummer_cols = {'[S1]': 'reference_start',\n",
    "               '[E1]': 'reference_end',\n",
    "              '[S2]': 'query_start',\n",
    "              '[E2]': 'query_end',\n",
    "              '[LEN 1]': 'align_length_ref',\n",
    "              '[LEN 2]': 'align_length_query',\n",
    "              '[LEN R]': 'reference_length',\n",
    "              '[LEN Q]': 'query_length',\n",
    "              '[% IDY]': 'pct_identity',\n",
    "              '[TAGS]': 'reference_id',\n",
    "              '[QID]': 'query_id'}\n",
    "\n",
    "# such a good tool and yet such an outdated output formatting...\n",
    "# this makes me sad\n",
    "mummer_header = '[S1]\\t[E1]\\t[S2]\\t[E2]\\t[LEN 1]\\t[LEN 2]\\t[% IDY]\\t[LEN R]\\t[LEN Q]\\t[TAGS]\\t[QID]'.split('\\t')\n",
    "\n",
    "title_filetype = {'cmp': 'Raw',\n",
    "                 'tag-h1': 'Hap1',\n",
    "                 'tag-h2': 'Hap2',\n",
    "                 'tag-un': 'Untag',\n",
    "                 'tag-h1-un': 'Hap1-Un',\n",
    "                 'tag-h2-un': 'Hap2-Un'}\n",
    "\n",
    "prettify_sample = {\n",
    "    'HG00514.sequel2.pb-ccs': 'HG.514 PB-Sq2 CCS',\n",
    "    'HG00514.sequel2.pb-clr': 'HG.514 PB-Sq2 CLR',\n",
    "    'HG00514.sequel2.pb-clr-25': 'HG.514 PB-Sq2 CLR (0.25)'\n",
    "}\n",
    "\n",
    "inpath = '/MMCI/TM/scratch/pebert/diploid_assembly/output/assembly_analysis/mummer'\n",
    "outpath = '/MMCI/TM/scratch/pebert/diploid_assembly/output/plots/alignment_ideograms'\n",
    "os.makedirs(outpath, exist_ok=True)\n",
    "coord_files = [os.path.join(inpath, f) for f in os.listdir(inpath) if f.endswith('coord.tab')]\n",
    "\n",
    "for table in coord_files:\n",
    "    if '_vs_' in table:\n",
    "        continue\n",
    "    output = os.path.join(outpath, os.path.basename(table).replace('coord.tab', 'aln_ideo.png'))\n",
    "    if os.path.isfile(output):\n",
    "        continue\n",
    "    parts = os.path.basename(table).split('.')\n",
    "    sample = '.'.join(parts[:2])\n",
    "    mobj = re.search('tag[\\-h12un]+', os.path.basename(table))\n",
    "    assert mobj is not None, 'No haplotype detected'\n",
    "    filetype = mobj.group(0)\n",
    "    filetype = title_filetype[filetype]\n",
    "    \n",
    "    filename = os.path.basename(table)\n",
    "    file_title = filename.replace('sequel2', 'sq2').replace('HG00514', 'HG514').replace('.coord.tab', '')\n",
    "    file_title = file_title.replace('_to_', '\\nmap to\\n', 1)\n",
    "    file_title = file_title.replace('pb-clr-25', 'CLR25').replace('pb-clr', 'CLR')\n",
    "    file_title = file_title.replace('pb-ccs', 'CCS')\n",
    "    \n",
    "    # deliberately skip the header row as the query ID\n",
    "    # is not part of it for whatever reason\n",
    "    aln = pandas.read_table(table, skiprows=3+1, sep='\\t',\n",
    "                            index_col=False, names=mummer_header)\n",
    "    new_cols = [mummer_cols[c] for c in aln.columns]\n",
    "    aln.columns = new_cols\n",
    "    aln['chrom'] = aln['reference_id']\n",
    "    aln = aln[aln.chrom.apply(lambda x: x in chromosome_list)].copy()\n",
    "    aln['start'] = aln['reference_start']\n",
    "    aln['width'] = aln['align_length_ref']\n",
    "    aln['colors'] = 'white'\n",
    "    # better safe than sorry\n",
    "    aln.sort_values(['chrom', 'start'], inplace=True)\n",
    "    aln.reset_index(drop=True, inplace=True)\n",
    "    if False:\n",
    "        # simple alternating color scheme\n",
    "        aln.loc[list(range(0, aln.shape[0], 2)), ['colors']] = 'dodgerblue'\n",
    "        aln.loc[list(range(1, aln.shape[0], 2)), ['colors']] = 'orange'\n",
    "    else:\n",
    "        last_query_id = aln.at[0, 'query_id']\n",
    "        last_chrom = aln.at[0, 'chrom']\n",
    "        color_scheme = []\n",
    "        c1, c2 = 'dodgerblue', 'orange'\n",
    "        for row in aln.itertuples():\n",
    "            if row.chrom != last_chrom:\n",
    "                c1, c2 = 'dodgerblue', 'orange'\n",
    "                last_chrom = row.chrom\n",
    "                last_query_id = row.query_id\n",
    "            if row.query_id != last_query_id:\n",
    "                last_query_id = row.query_id\n",
    "                c1, c2 = c2, c1\n",
    "            color_scheme.append(c1)\n",
    "        aln['colors'] = color_scheme\n",
    "\n",
    "    fig = plt.figure(figsize=figsize)\n",
    "    ax = fig.add_subplot(111)\n",
    "\n",
    "    # Now all we have to do is call our function for the ideogram data...\n",
    "    #print(\"adding ideograms...\")\n",
    "    for collection in chromosome_collections(ideo, chrom_ybase, chrom_height, edgecolor='black'):\n",
    "        ax.add_collection(collection)\n",
    "\n",
    "    # ...and the gene data\n",
    "    #print(\"adding alignments...\")\n",
    "    for collection in chromosome_collections(\n",
    "        aln, gene_ybase, gene_height, linewidths=0\n",
    "    ):\n",
    "        ax.add_collection(collection)\n",
    "\n",
    "    text = 'Alignments: {}\\n'.format(aln.shape[0])\n",
    "    text += 'Median length: {} kbp\\n'.format(int(round(aln['width'].median() / 1e3, 0)))\n",
    "    text += 'Coverage: ~{}x'.format(round(aln['width'].sum() / (3 * 1e9), 2))\n",
    "    \n",
    "    txt = ax.text(0.6, 0.2, s=text, transform=ax.transAxes, fontdict={'fontsize': 14})\n",
    "        \n",
    "    # Axes tweaking\n",
    "    ax.set_yticks([chrom_centers[i] for i in chromosome_list])\n",
    "    ax.set_yticklabels(chromosome_list)\n",
    "    ax.axis('tight')\n",
    "    tt = ax.set_title('{}: {}'.format(filetype, file_title), fontsize=12)\n",
    "    tt.set_position([0.5, 1.025])\n",
    "    ax.set_xlabel('Genomic coordinate', fontsize=12)\n",
    "    ax.set_ylabel('Chromosome', fontsize=14)\n",
    "    \n",
    "    fig.savefig(output, dpi=600, bbox_inches='tight', extra_artists=[txt])\n",
    "    plt.close(fig)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
