{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CLR high   65.69\n",
      "any high   69.0\n",
      "CLR low   0.62\n",
      "any low   0.39\n",
      "CLR medium   31.74\n",
      "any medium   28.48\n",
      "HiFi high   79.03\n",
      "HiFi low   0.3\n",
      "HiFi medium   20.37\n",
      "CLR   2778884550.5\n",
      "any   2785558664.0\n",
      "HiFi   2790407215.5\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import collections as col\n",
    "\n",
    "\"\"\"\n",
    "What does this do?\n",
    "Processes the output of \"process_bng_hybrid.py\" after aligning scaffolds\n",
    "back to reference assembly. This notebook computes the approximate scaffold\n",
    "placement and assigns the confidence score for the scaffolded contig alignments.\n",
    "\n",
    "This process is a prerequisite for computing the \"coverage mask\" (accessibility)\n",
    "based on the scaffolded contig alignments (as opposed to unscaffolded contig\n",
    "alignments, after MAPQ thresholding).\n",
    "\n",
    "Prints stats about pct. of high/medium confidence scaffolded contig alignments.\n",
    "\"\"\"\n",
    "\n",
    "scaffolds_path = '/home/local/work/data/hgsvc/figSX_panels/bng_scaffolds/v12'\n",
    "scfalign_path = '/home/local/work/data/hgsvc/figSX_panels/bng_scfaling/v12'\n",
    "cache_file = '/home/local/work/data/hgsvc/figSX_panels/bng_scfaling/cache.h5'\n",
    "\n",
    "print_stats = True\n",
    "\n",
    "align_columns = [\n",
    "    'chrom',\n",
    "    'start',\n",
    "    'end',\n",
    "    'name',\n",
    "    'mapq',\n",
    "    'orientation'\n",
    "]\n",
    "\n",
    "def load_scaffolds(assembly):\n",
    "    \n",
    "    pattern = os.path.join(scaffolds_path, '{}*.tsv'.format(assembly))\n",
    "    scf_file = glob.glob(pattern)\n",
    "    assert len(scf_file) == 1, 'Mulit scaffold match'\n",
    "    df = pd.read_csv(scf_file[0], sep='\\t', header=0, index_col=None)\n",
    "    return df\n",
    "\n",
    "\n",
    "def extract_contig_length(name_entry):\n",
    "    \n",
    "    ctg_info = name_entry.split('@')[-1]\n",
    "    positions = ctg_info.split(':')[-1]\n",
    "    start, end = positions.split('-')\n",
    "    return int(end) - int(start)\n",
    "\n",
    "\n",
    "def load_alignments(align_file):\n",
    "    \n",
    "    align_path = os.path.join(scfalign_path, align_file)\n",
    "    df = pd.read_csv(align_path, sep='\\t', names=align_columns, header=None)\n",
    "    df = df.loc[df['name'] != 'dummy', :].copy()\n",
    "    if df.empty:\n",
    "        return df\n",
    "\n",
    "    df['scaffold'] = df['name'].apply(lambda x: x.split('@')[0])\n",
    "    df['order_number'] = df['name'].apply(lambda x: int(x.split('@')[2]))\n",
    "    df['aligned_length'] = (df['end'] - df['start']).astype('int32')\n",
    "    df['midpoint'] = df['start'] + df['aligned_length'] // 2\n",
    "    df['contig_length'] = df['name'].apply(extract_contig_length)\n",
    "    df.sort_values(['chrom', 'start', 'end'], axis=0, ascending=True, inplace=True)\n",
    "    return df\n",
    "\n",
    "\n",
    "def compute_scaffold_midpoint_distances(scaffold, scaffold_layouts):\n",
    "    \n",
    "    select_scaffold = scaffold_layouts['object'] == scaffold\n",
    "    select_sequence = scaffold_layouts['component'] == 'sequence'\n",
    "    # scaffold length includes also assembly gaps\n",
    "    scaffold_length = int(scaffold_layouts.loc[scaffold_layouts['name'] == scaffold, 'length'])\n",
    "    \n",
    "    this_scf = scaffold_layouts.loc[select_scaffold & select_sequence, :].copy()\n",
    "    this_scf['midpoint'] = this_scf['start'] + this_scf['length'] // 2\n",
    "    weighted_midpoint = int(np.average(this_scf['midpoint'], weights=this_scf['length']))\n",
    "    this_scf['midpoint_dist'] = this_scf['midpoint'] - weighted_midpoint\n",
    "    \n",
    "    sequence_lengths = dict((row['name'], row['length']) for idx, row in this_scf.iterrows())\n",
    "    sequence_lengths['scaffold'] = scaffold_length\n",
    "    sequence_lengths['scaffold_no_gaps'] = sum([v for k, v in sequence_lengths.items() if k != 'scaffold'])\n",
    "\n",
    "    return this_scf, weighted_midpoint, sequence_lengths\n",
    "\n",
    "\n",
    "def compute_covered_sequence(alignments, midpoint, left_margin, right_margin):\n",
    "    \n",
    "    lower_bound = midpoint - left_margin\n",
    "    upper_bound = midpoint + right_margin\n",
    "    \n",
    "    select_left = alignments['end'] > lower_bound\n",
    "    select_right = alignments['start'] < upper_bound\n",
    "    \n",
    "    total_overlap = 0\n",
    "    indices = []\n",
    "    for idx, row in alignments.loc[select_left & select_right, :].iterrows():\n",
    "        indices.append(idx)\n",
    "        overlap = min(upper_bound, row['end']) - max(lower_bound, row['start'])\n",
    "        assert overlap > 0, 'Something is wrong'\n",
    "        total_overlap += overlap\n",
    "    if total_overlap == 0:\n",
    "        raise ValueError('No alignments selected: {} to {}\\n\\n{}'.format(lower_bound, upper_bound, alignments))\n",
    "    \n",
    "    return total_overlap, indices\n",
    "\n",
    "\n",
    "def compute_alignment_midpoint_distances(contig_alignments):\n",
    "    \n",
    "    try:\n",
    "        aln_wt_midpoint = int(np.average(\n",
    "            contig_alignments['midpoint'],\n",
    "            weights=contig_alignments['aligned_length']\n",
    "        ))\n",
    "    except ZeroDivisionError:\n",
    "        print(contig_alignments)\n",
    "        raise\n",
    "    \n",
    "    aln_midpoints = col.defaultdict(dict)\n",
    "    for idx, rows in contig_alignments.groupby('name'):\n",
    "        wt_midpoint = int(np.average(\n",
    "            rows['midpoint'],\n",
    "            weights=rows['aligned_length']\n",
    "        ))\n",
    "        # Super-Scaffold_363@chr17@1@frw@cluster12_scaffold_95@ctg:0-524872\n",
    "        contig_name = idx.split('@')[-2]\n",
    "        aln_midpoints[contig_name]['midpoint'] = wt_midpoint\n",
    "        aln_midpoints[contig_name]['midpoint_dist'] = wt_midpoint - aln_wt_midpoint\n",
    "\n",
    "    return aln_midpoints\n",
    "\n",
    "\n",
    "def check_overlaps(contig_alignments, aln_wt_midpoint, left_margin, right_margin):\n",
    "\n",
    "    # since we don't know the relative orientation of the scaffold,\n",
    "    # check which selection covers more aligned sequence\n",
    "\n",
    "    is_reverse = False\n",
    "    \n",
    "    forward_overlaps, forward_indices = compute_covered_sequence(\n",
    "        contig_alignments,\n",
    "        aln_wt_midpoint,\n",
    "        left_margin,\n",
    "        right_margin\n",
    "    )\n",
    "    \n",
    "    reverse_overlaps, reverse_indices = compute_covered_sequence(\n",
    "        contig_alignments,\n",
    "        aln_wt_midpoint,\n",
    "        right_margin,\n",
    "        left_margin\n",
    "    )\n",
    "\n",
    "    if forward_overlaps == 0 and reverse_overlaps == 0:\n",
    "        raise ValueError('No overlaps')\n",
    "\n",
    "    if forward_overlaps > reverse_overlaps:\n",
    "        selected_alignments = forward_indices\n",
    "    elif forward_overlaps < forward_overlaps:\n",
    "        selected_alignments = reverse_indices\n",
    "        is_reverse = True\n",
    "    else:\n",
    "        # seems not to matter\n",
    "        selected_alignments = forward_indices\n",
    "    return selected_alignments, is_reverse\n",
    "\n",
    "\n",
    "def exclude_isolated_blocks(contig_alignments):\n",
    "    \n",
    "    # compute all vs all midpoint distances\n",
    "    avg_distances = []\n",
    "    for idx, row in contig_alignments.iterrows():\n",
    "        \n",
    "        not_self = contig_alignments.index != idx\n",
    "        avg_dist = (contig_alignments.loc[not_self, 'midpoint'] - row['midpoint']).abs().sum() / not_self.sum()\n",
    "        avg_distances.append((avg_dist, idx))\n",
    "        \n",
    "    avg_distances = sorted(avg_distances, reverse=True)\n",
    "    exclude_indices = set()\n",
    "    for pos, (d, i) in enumerate(avg_distances):\n",
    "        try:\n",
    "            if d > avg_distances[pos+1][0]:\n",
    "                exclude_indices.add(i)\n",
    "                break\n",
    "        except IndexError:\n",
    "            break\n",
    "        exclude_indices.add(i)\n",
    "\n",
    "    scattered = contig_alignments.index.isin(exclude_indices)\n",
    "        \n",
    "    aln_midpoint = int(np.average(\n",
    "        contig_alignments.loc[~scattered, 'midpoint'],\n",
    "        weights=contig_alignments.loc[~scattered, 'aligned_length']\n",
    "    ))\n",
    "\n",
    "    return aln_midpoint\n",
    "\n",
    "\n",
    "def select_alignments_for_placement(\n",
    "    contig_alignments,\n",
    "    seq_lens,\n",
    "    scf_left_margin, \n",
    "    scf_right_margin, \n",
    "    is_single_split):\n",
    "    \n",
    "    # compute weighted midpoint for all alignments\n",
    "    aln_midpoint = int(np.average(\n",
    "        contig_alignments['midpoint'],\n",
    "        weights=contig_alignments['aligned_length']\n",
    "    ))\n",
    "\n",
    "    # allow some slack for potential variants\n",
    "    left_margin = int(scf_left_margin + scf_left_margin * 0.05)\n",
    "    right_margin = int(scf_right_margin + scf_right_margin * 0.05)\n",
    "        \n",
    "    try:\n",
    "        selected_alignments, is_reverse = check_overlaps(\n",
    "            contig_alignments,\n",
    "            aln_midpoint,\n",
    "            left_margin,\n",
    "            right_margin\n",
    "        )\n",
    "    except ValueError:\n",
    "        if is_single_split:\n",
    "            # reset midpoint to best aligned block\n",
    "            best_alignment = contig_alignments['aligned_length'] * contig_alignments['mapq']\n",
    "            best_index = best_alignment.index[best_alignment == best_alignment.max()]\n",
    "            aln_midpoint = int(contig_alignments.loc[best_index, 'start']) + int(contig_alignments.loc[best_index, 'aligned_length'] // 2)\n",
    "            selected_alignments, is_reverse = check_overlaps(\n",
    "                contig_alignments,\n",
    "                aln_midpoint,\n",
    "                left_margin,\n",
    "                right_margin\n",
    "            )\n",
    "        elif contig_alignments.shape[0] < 3:\n",
    "            # not enough aligned contigs to make an attempt to save this\n",
    "            # scaffold\n",
    "            raise\n",
    "        else:\n",
    "            # make one attempt to drop the one alignment (or multiple if they all\n",
    "            # have the same midpoint; may happen for low-scoring fragments) that\n",
    "            # has largest average distance to all other alignments;\n",
    "            # if alignments are too scattered,\n",
    "            # this won't change anything and still result in low confidence\n",
    "            aln_midpoint = exclude_isolated_blocks(contig_alignments)\n",
    "            selected_alignments, is_reverse = check_overlaps(\n",
    "                contig_alignments,\n",
    "                aln_midpoint,\n",
    "                left_margin,\n",
    "                right_margin\n",
    "            )\n",
    "    \n",
    "    # compute alignment midpoint distances\n",
    "    aln_midpoint_dists = compute_alignment_midpoint_distances(\n",
    "        contig_alignments.loc[selected_alignments, :].copy()\n",
    "    )\n",
    "    \n",
    "    # set the boundaries of the scaffold to the maximum extension\n",
    "    min_start = contig_alignments.loc[selected_alignments, 'start'].min()\n",
    "    max_end = contig_alignments.loc[selected_alignments, 'end'].max()\n",
    "    \n",
    "    aligned_length = contig_alignments.loc[selected_alignments, 'aligned_length'].sum()\n",
    "    # check that the contig is not placed based on a marginal amount of sequence\n",
    "    # unlikely to happen, but better safe than sorry...\n",
    "    scf_length_nogap = seq_lens['scaffold_no_gaps']\n",
    "    if aligned_length < scf_length_nogap * 0.5:\n",
    "        raise ValueError('Insufficient aligned length for placing scaffold')\n",
    "    \n",
    "    if is_reverse:\n",
    "        scf_left_margin, scf_right_margin = scf_right_margin, scf_left_margin\n",
    "    \n",
    "    scf_midpoint = aln_midpoint\n",
    "    \n",
    "    scf_left_margin = min(min_start, scf_midpoint - scf_left_margin)\n",
    "    scf_right_margin = max(max_end, scf_midpoint - scf_right_margin)\n",
    "    \n",
    "    return selected_alignments, aln_midpoint_dists, scf_midpoint, scf_left_margin, scf_right_margin\n",
    " \n",
    "\n",
    "def compute_displacement_stats(scaffold, scaffold_length, aln_midpoint_dists):\n",
    "    # since the scaffold placement does not take structural variation into\n",
    "    # account, omit this for now - can be quite off\n",
    "    raise RuntimeError('Do not use this funtion')\n",
    "    contig_displacements = []\n",
    "    for idx, contig in scaffold.iterrows():\n",
    "        ctg = contig['name']\n",
    "        if ctg not in aln_midpoint_dists:\n",
    "            continue\n",
    "        delta = abs(abs(contig['midpoint_dist']) - abs(aln_midpoint_dists[ctg]['midpoint_dist']))\n",
    "        contig_displacements.append(delta)\n",
    "\n",
    "    contig_displacements = np.array(contig_displacements, dtype=np.int32)\n",
    "    total_displacement = contig_displacements.sum()\n",
    "    rel_displacement = (total_displacement / scf_length * 100).round(2) \n",
    "    return total_displacement, rel_displacement\n",
    "\n",
    "\n",
    "def determine_scattered_or_single_split(alignments, select_scf_aln):\n",
    "\n",
    "    # if scattered over multiple chromosomes, assign all low confidence\n",
    "    is_scattered = alignments.loc[select_scf_aln, 'chrom'].nunique() > 1\n",
    "    if is_scattered:\n",
    "        alignments.loc[select_scf_aln, 'loc_confidence'] = 'low'\n",
    "        alignments.loc[select_scf_aln, 'scaffold_strand'] = '.'\n",
    "\n",
    "    # if single contig scaffold, opt for assigning high confidence\n",
    "    is_single_contig = alignments.loc[select_scf_aln, 'chrom'].shape[0] == 1\n",
    "    if is_single_contig:\n",
    "        alignments.loc[select_scf_aln, 'loc_confidence'] = 'high'\n",
    "        orientation = alignments.loc[select_scf_aln, 'orientation'].values[0]\n",
    "        alignments.loc[select_scf_aln, 'scaffold_strand'] = orientation\n",
    "\n",
    "    # single split contig\n",
    "    # example case: HG00096 / H1 / Super-Scaffold_101491\n",
    "    # cen or peri-cen, alignment scattered, HET region on chr1q12\n",
    "    num_contigs = alignments.loc[select_scf_aln, 'name'].nunique()\n",
    "    num_chroms = alignments.loc[select_scf_aln, 'chrom'].nunique()\n",
    "    is_single_split_contig = num_contigs == 1 and num_chroms == 1\n",
    "\n",
    "    return is_scattered, is_single_contig, is_single_split_contig\n",
    "    \n",
    "    \n",
    "def determine_scaffold_placement_confidence(alignments, scaffolds):\n",
    "    \n",
    "    alignments['loc_confidence'] = 'undetermined'\n",
    "    alignments['scaffold_strand'] = 'undetermined'\n",
    "    alignments['scaffold_start'] = -1\n",
    "    alignments['scaffold_end'] = -1\n",
    "        \n",
    "    for scf in set(alignments['scaffold'].values):\n",
    "        \n",
    "        select_scf_aln = alignments['scaffold'] == scf\n",
    "        \n",
    "        is_scattered, is_single, is_single_split_contig = determine_scattered_or_single_split(\n",
    "            alignments,\n",
    "            select_scf_aln\n",
    "        )\n",
    "        if is_scattered:\n",
    "            # side effect: alignments set to low in call above\n",
    "            continue\n",
    "        if is_single and not is_single_split_contig:\n",
    "            # side effect: alignment set to high in call above\n",
    "            continue\n",
    "        \n",
    "        # complicated part: heuristic to determine approximate placement\n",
    "        # of scaffold relative to hg38\n",
    "        scaffold, scf_midpoint, seq_lengths = compute_scaffold_midpoint_distances(scf, scaffolds)\n",
    "        \n",
    "        # left margin: 0 ... MID\n",
    "        left_margin = scf_midpoint - 1\n",
    "        # right margin: MID ... END\n",
    "        right_margin = seq_lengths['scaffold'] - scf_midpoint\n",
    "        \n",
    "        contig_aln = alignments.loc[select_scf_aln, :].copy()\n",
    "        try:\n",
    "            selected_aln, aln_midpoint_dists, scf_midpoint, left_margin, right_margin = select_alignments_for_placement(\n",
    "                contig_aln,\n",
    "                seq_lengths,\n",
    "                left_margin,\n",
    "                right_margin,\n",
    "                is_single_split_contig\n",
    "            )\n",
    "        except ValueError:\n",
    "            # impossible to find good placement, set everything to low confidence\n",
    "            # example: Super-Scaffold_101620 / HG00096 H1 scattered on chr4\n",
    "            alignments.loc[select_scf_aln, 'loc_confidence'] = 'low'\n",
    "            alignments.loc[select_scf_aln, 'scaffold_strand'] = '.'\n",
    "            continue\n",
    "        \n",
    "        aligned_order = contig_aln.loc[selected_aln, 'order_number'].drop_duplicates(keep='first', inplace=False)\n",
    "                \n",
    "        # note that this implies that all contigs\n",
    "        # have been aligned/used, although not necessarily\n",
    "        # to 100% of their size\n",
    "        if np.array_equal(aligned_order.values, scaffold['order'].values):\n",
    "            # scaffold orientation -> forward\n",
    "            alignments.loc[selected_aln, 'loc_confidence'] = 'high'\n",
    "            alignments.loc[selected_aln, 'scaffold_strand'] = '+'\n",
    "        elif np.array_equal(aligned_order.values, np.flip(scaffold['order'].values)):\n",
    "            # scaffold orientation -> reverse\n",
    "            alignments.loc[selected_aln, 'loc_confidence'] = 'high'\n",
    "            alignments.loc[selected_aln, 'scaffold_strand'] = '-'\n",
    "        else:\n",
    "            majority_orientation = contig_aln.loc[selected_aln, 'orientation'].value_counts().sort_values(inplace=False, ascending=False)\n",
    "            majority_orientation = majority_orientation.index[0]\n",
    "            alignments.loc[selected_aln, 'loc_confidence'] = 'medium'\n",
    "            alignments.loc[selected_aln, 'scaffold_strand'] = majority_orientation\n",
    "        \n",
    "        alignments.loc[selected_aln, 'scaffold_start'] = scf_midpoint - left_margin\n",
    "        alignments.loc[selected_aln, 'scaffold_end'] = scf_midpoint + right_margin\n",
    "        \n",
    "        # in case contigs have not been used for scaffold placement\n",
    "        unused_contigs = contig_aln.index[~contig_aln.index.isin(selected_aln)]\n",
    "        if not unused_contigs.empty:\n",
    "            alignments.loc[unused_contigs, 'loc_confidence'] = 'low'\n",
    "            alignments.loc[unused_contigs, 'scaffold_strand'] = '.'\n",
    "\n",
    "    if (alignments['loc_confidence'] == 'undetermined').any():\n",
    "        raise RuntimeError('Undetermined alignment confidence')\n",
    "            \n",
    "    return alignments\n",
    "\n",
    "\n",
    "\n",
    "if not os.path.isfile(cache_file):\n",
    "\n",
    "    collector = col.defaultdict(list)\n",
    "\n",
    "    for scfalign in sorted(os.listdir(scfalign_path)):\n",
    "        if not scfalign.endswith('.bed'):\n",
    "            continue\n",
    "        assembly = scfalign.split('_map-to_')[0]\n",
    "        scaffolds = load_scaffolds(assembly)\n",
    "        alignments = load_alignments(scfalign)\n",
    "        if alignments.empty:\n",
    "            continue\n",
    "        alignments = determine_scaffold_placement_confidence(alignments, scaffolds)\n",
    "\n",
    "        collector[assembly].append(alignments)\n",
    "\n",
    "    with pd.HDFStore(cache_file, 'w', complevel=5) as hdf:\n",
    "        for k, v in collector.items():\n",
    "            merged = pd.concat(v, axis=0, ignore_index=False)\n",
    "            merged.sort_values(['chrom', 'start', 'end'], axis=0, inplace=True)\n",
    "            hdf.put(k.replace('.', '_').replace('-', ''), merged, format='fixed')\n",
    "\n",
    "\n",
    "out_path = '/home/local/work/data/hgsvc/figSX_panels/bng_scfaling/processed'\n",
    "            \n",
    "aligned_length = col.defaultdict(col.Counter)\n",
    "score_map = {\n",
    "    'high': 1000,\n",
    "    'medium': 500,\n",
    "    'low': 0\n",
    "}\n",
    "with pd.HDFStore(cache_file, 'r') as hdf:\n",
    "    for k in hdf.keys():\n",
    "        sample = k.split('_')[0].strip('/')\n",
    "        if 'h1un' in k:\n",
    "            hap = 'H1'\n",
    "        else:\n",
    "            hap = 'H2'\n",
    "        if 'pbsq2ccs' in k:\n",
    "            tech = 'HiFi'\n",
    "        else:\n",
    "            tech = 'CLR'\n",
    "        df = hdf[k]\n",
    "        df['score'] = df['loc_confidence'].apply(lambda x: score_map[x])\n",
    "        bed = df[['chrom', 'start', 'end', 'name', 'score', 'orientation']].copy()\n",
    "        bed.sort_values(['chrom', 'start', 'end'], axis=0, inplace=True)\n",
    "        if 'h1un' in k:\n",
    "            dump_file = k.strip('/').replace('_h1un_', '.h1-un.').replace('pbsq2', 'pbsq2-')\n",
    "        else:\n",
    "            dump_file = k.strip('/').replace('_h2un_', '.h2-un.').replace('pbsq2', 'pbsq2-')\n",
    "        dump_file += '_map-to_GRCh38.scf-conf.bed'\n",
    "        dump_path = os.path.join(out_path, dump_file)\n",
    "\n",
    "        with open(dump_path, 'w') as dump:\n",
    "            _ = dump.write('#')\n",
    "            bed.to_csv(dump, header=True, index=False, sep='\\t')\n",
    "        \n",
    "        if print_stats:\n",
    "            grouped = df.groupby('loc_confidence')['aligned_length'].sum()\n",
    "            for k, v in grouped.items():\n",
    "                aligned_length[(sample, tech, hap)][k] += v\n",
    "    \n",
    "collector = col.defaultdict(list)\n",
    "total_aligned = col.defaultdict(list)\n",
    "# compute per sample stats\n",
    "for (s, t, h), counts in aligned_length.items():\n",
    "    total = sum(counts.values())\n",
    "    total_aligned[t].append(total)\n",
    "    total_aligned['any'].append(total)\n",
    "    for loc_conf, count in counts.items():\n",
    "        rel = round(count / total * 100, 2)\n",
    "        collector[(t, loc_conf)].append(rel)\n",
    "        collector[('any', loc_conf)].append(rel)\n",
    "\n",
    "if print_stats:\n",
    "    for (tech, loc_conf), stats in collector.items():\n",
    "        print(tech, loc_conf, ' ', np.median(stats).round(2))\n",
    "    for tech, total in total_aligned.items():\n",
    "        print(tech, ' ', np.median(total).round(2))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
