{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "################\n",
      "==== 3q29\n",
      "################\n",
      "################\n",
      "==== HLA\n",
      "################\n",
      "################\n",
      "==== HLA-F\n",
      "################\n",
      "################\n",
      "==== HLA-G\n",
      "################\n",
      "################\n",
      "==== HLA-A\n",
      "################\n",
      "################\n",
      "==== HLA-E\n",
      "################\n",
      "################\n",
      "==== HLA-C\n",
      "################\n",
      "################\n",
      "==== HLA-B\n",
      "################\n",
      "################\n",
      "==== HLA-DRA\n",
      "################\n",
      "################\n",
      "==== HLA-DRB5\n",
      "################\n",
      "################\n",
      "==== HLA-DRB1\n",
      "################\n",
      "################\n",
      "==== HLA-DQA1\n",
      "################\n",
      "################\n",
      "==== HLA-DQB1\n",
      "################\n",
      "################\n",
      "==== HLA-DQA2\n",
      "################\n",
      "################\n",
      "==== HLA-DQB2\n",
      "################\n",
      "################\n",
      "==== HLA-DOB\n",
      "################\n",
      "################\n",
      "==== HLA-DMB\n",
      "################\n",
      "################\n",
      "==== HLA-DMA\n",
      "################\n",
      "################\n",
      "==== HLA-DOA\n",
      "################\n",
      "################\n",
      "==== HLA-DPA1\n",
      "################\n",
      "################\n",
      "==== HLA-DPB1\n",
      "################\n",
      "          complete  partial  missing\n",
      "HLA-A           88        0        0\n",
      "HLA-B           88        0        0\n",
      "HLA-C           88        0        0\n",
      "HLA-DMA         88        0        0\n",
      "HLA-DMB         88        0        0\n",
      "HLA-DOA         88        0        0\n",
      "HLA-DOB         88        0        0\n",
      "HLA-DPA1        88        0        0\n",
      "HLA-DPB1        88        0        0\n",
      "HLA-DQA1        88        0        0\n",
      "HLA-DQA2        88        0        0\n",
      "HLA-DQB1        88        0        0\n",
      "HLA-DQB2        88        0        0\n",
      "HLA-DRA         88        0        0\n",
      "HLA-DRB1        86        1        1\n",
      "HLA-DRB5        41        4       43\n",
      "HLA-E           88        0        0\n",
      "HLA-F           88        0        0\n",
      "HLA-G           88        0        0\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os as os\n",
    "import io\n",
    "import collections as col\n",
    "import numpy\n",
    "import operator as op\n",
    "import numpy as np\n",
    "import numpy.ma as ma\n",
    "import importlib\n",
    "\n",
    "\"\"\"\n",
    "What does this do?\n",
    "Read the output of a \"bedtools intersect\" operation for all haploid assemblies\n",
    "(MAPQ60) and the ROI file (e.g., HLA, 3q29), see \"recomp_cov.py\" script lines ~66ff\n",
    "Extract assembly infos and compute statistics:\n",
    "- number of contigs\n",
    "- size/number of gaps\n",
    "- coverage (bp, pct)\n",
    "etc.\n",
    "Results are used for creating Supp. Table(s); produces also output for condensed\n",
    "coverage plots (so far not used in manuscript), see below\n",
    "\"\"\"\n",
    "\n",
    "\"\"\"\n",
    "Input / Output\n",
    "\n",
    "# /home/local/work/data/hgsvc/roi/MAPQ-ALL_roi_assm_overlaps.tsv\n",
    "# Update:\n",
    "/home/local/work/data/hgsvc/roi/MAPQ-ALL_roi_assm_HLAn_overlaps.tsv\n",
    "\n",
    "/home/local/work/data/hgsvc/roi/roi_HLA_3q29_blocks.tsv\n",
    "/home/local/work/data/hgsvc/roi/roi_assembly_summary_{ROI}.tsv\n",
    "\"\"\"\n",
    "\n",
    "plot_aux_module = '/home/local/work/code/github/project-diploid-assembly/notebooks/aux_mods/plot_aux.py'\n",
    "plot_aux_spec = importlib.util.spec_from_file_location(\"plot_aux\", plot_aux_module)\n",
    "plot_aux = importlib.util.module_from_spec(plot_aux_spec)\n",
    "plot_aux_spec.loader.exec_module(plot_aux)\n",
    "\n",
    "ann_path = '/home/local/work/code/github/project-diploid-assembly/annotation'\n",
    "st = os.path.join(ann_path, 'sample_table.tsv')\n",
    "pt = os.path.join(ann_path, '1kg_hgsvc_colors.csv')\n",
    "samples = plot_aux.load_sample_table(st)\n",
    "hexcodes, rgbcodes, popmap = plot_aux.load_population_annotation(pt)\n",
    "\n",
    "prop = plot_aux.get_plot_property\n",
    "\n",
    "tsv_file = \"/home/local/work/data/hgsvc/roi/MAPQ-ALL_roi_HLAn_assm_overlaps.tsv\"\n",
    "\n",
    "tech_map = {\n",
    "    'pbsq2-clr': 'CLR',\n",
    "    'pbsq2-ccs': 'HiFi'\n",
    "}\n",
    "\n",
    "hap_map = {\n",
    "    'h1-un': 'H1',\n",
    "    'h2-un': 'H2'\n",
    "}\n",
    "\n",
    "def extract_assembly_info(assembly):\n",
    "    \n",
    "    filename = os.path.basename(assembly)\n",
    "    parts = filename.split('_')\n",
    "    sample = parts[0]\n",
    "    tech = tech_map[parts[2]]\n",
    "    hap = hap_map[parts[3].split('.')[1]]\n",
    "    return sample, tech, hap\n",
    "    \n",
    "\n",
    "names = [\n",
    "    'roi_chrom',\n",
    "    'roi_start',\n",
    "    'roi_end',\n",
    "    'roi',\n",
    "    'assembly',\n",
    "    'qry_chrom',\n",
    "    'qry_start',\n",
    "    'qry_end',\n",
    "    'contig_name',\n",
    "    'mapq',\n",
    "    'strand',\n",
    "    'overlap'\n",
    "]\n",
    "\n",
    "df = pd.read_csv(tsv_file, sep='\\t', header=None, names=names)\n",
    "df = df.loc[df['mapq'] == 60, :].copy()\n",
    "df.reset_index(drop=True, inplace=True)\n",
    "df['roi_length'] = df['roi_end'] - df['roi_start']\n",
    "\n",
    "assm_info = df['assembly'].apply(extract_assembly_info)\n",
    "df2 = pd.DataFrame.from_records(\n",
    "    assm_info,\n",
    "    columns=['sample', 'platform', 'haplotype'],\n",
    "    index=df.index\n",
    ")\n",
    "\n",
    "df = pd.concat([df, df2], ignore_index=False, axis=1)\n",
    "df.drop(['assembly'], axis=1, inplace=True)\n",
    "\n",
    "get_index_items = op.itemgetter(*tuple([12, 13, 14, 3]))\n",
    "\n",
    "df.index = pd.MultiIndex.from_tuples(\n",
    "    [get_index_items(t) for t in df.itertuples(index=None, name=None)],\n",
    "    names=['sample', 'platform', 'haplotype', 'roi']\n",
    ")\n",
    "df.drop(['sample', 'platform', 'haplotype', 'roi'], axis=1, inplace=True)\n",
    "\n",
    "summed_ovl = df.groupby(['sample', 'platform', 'haplotype', 'roi'])['overlap'].sum()\n",
    "num_contigs = df.groupby(['sample', 'platform', 'haplotype', 'roi'])['contig_name'].nunique()\n",
    "\n",
    "joined = pd.concat([summed_ovl, num_contigs], axis=1, ignore_index=False)\n",
    "joined.rename({'overlap': 'overlap_bp', 'contig_name': 'contig_count'}, axis=1, inplace=True)\n",
    "\n",
    "joined = joined.join(df, how='outer')\n",
    "\n",
    "mrg_aligns = []\n",
    "mrg_index = []\n",
    "\n",
    "color_bars = []\n",
    "grey = (160/255, 160/255, 160/255)\n",
    "black = (0, 0, 0)\n",
    "\n",
    "for (sample, platform, hap, roi), aligns in joined.groupby(['sample', 'platform', 'haplotype', 'roi']):\n",
    "    roi_mask = np.zeros(aligns['roi_length'].values[0], dtype=np.bool)\n",
    "    cov_mask = np.zeros(aligns['roi_length'].values[0], dtype=np.int8)\n",
    "    \n",
    "    roi_start = aligns['roi_start'].values[0]\n",
    "    roi_end = aligns['roi_end'].values[0]\n",
    "    roi_length = roi_end - roi_start\n",
    "    roi_coords = np.arange(roi_start, roi_end)\n",
    "    contig_count = aligns['contig_count'].values[0]\n",
    "    \n",
    "    for idx, aln in aligns.iterrows():\n",
    "        start = max(roi_start, aln['qry_start'])\n",
    "        end = min(roi_end, aln['qry_end'])\n",
    "        roi_mask[start-roi_start:end-roi_start] |= True\n",
    "        cov_mask[start-roi_start:end-roi_start] += 1\n",
    "\n",
    "    coverage = roi_mask.sum()\n",
    "\n",
    "    # count gaps\n",
    "    gaps = np.array([\n",
    "        s.stop - s.start for s in \n",
    "        ma.clump_unmasked(ma.masked_array(roi_coords, mask=roi_mask))\n",
    "    ], dtype=np.int32)\n",
    "    num_gaps = gaps.size\n",
    "    if num_gaps > 0:\n",
    "        min_gap = gaps.min()\n",
    "        max_gap = gaps.max()\n",
    "        avg_gap = int(gaps.mean())\n",
    "    else:\n",
    "        min_gap = 0\n",
    "        max_gap = 0\n",
    "        avg_gap = 0\n",
    "        \n",
    "    sample_color = rgbcodes[samples[sample]['super_population']]\n",
    "    for cov, color in zip([0, 1, 2], [black, sample_color, grey]):\n",
    "        if cov == 2:\n",
    "            select_mask = cov_mask >= cov\n",
    "            block_type = 'multi'\n",
    "        else:\n",
    "            if cov == 0:\n",
    "                block_type = 'gap'\n",
    "            else:\n",
    "                block_type = 'align'\n",
    "            select_mask = cov_mask == cov\n",
    "        slices = ma.clump_masked(ma.masked_array(roi_coords, mask=select_mask))\n",
    "        for s in slices:\n",
    "            color_bars.append(\n",
    "                (\n",
    "                    sample,\n",
    "                    platform,\n",
    "                    hap,\n",
    "                    roi,\n",
    "                    coverage,\n",
    "                    round(coverage / roi_length, 3),\n",
    "                    contig_count,\n",
    "                    roi_coords[s.start],\n",
    "                    s.stop - s.start,\n",
    "                    color\n",
    "                )\n",
    "            )\n",
    "    \n",
    "    mrg_aligns.append(\n",
    "        (\n",
    "            coverage,\n",
    "            contig_count,\n",
    "            roi_length - coverage,\n",
    "            (coverage / roi_length * 100).round(2),\n",
    "            num_gaps,\n",
    "            min_gap,\n",
    "            max_gap,\n",
    "            avg_gap\n",
    "        )\n",
    "    )\n",
    "    mrg_index.append((sample, platform, hap, roi))\n",
    "\n",
    "aln_colors = pd.DataFrame(\n",
    "    color_bars,\n",
    "    columns=[\n",
    "        'sample',\n",
    "        'platform',\n",
    "        'hap',\n",
    "        'roi',\n",
    "        'coverage',\n",
    "        'cov_pct',\n",
    "        'contigs',\n",
    "        'block_start',\n",
    "        'block_length',\n",
    "        'block_color'\n",
    "    ]\n",
    ")\n",
    "\n",
    "# this output file can be used to create a condensed representation\n",
    "# of alignment coverage in the respective locus (colored by super pop)\n",
    "aln_colors.to_csv(\n",
    "    '/home/local/work/data/hgsvc/roi/roi_HLAn_3q29_blocks.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False\n",
    ")\n",
    "\n",
    "\n",
    "joined = pd.DataFrame(\n",
    "    mrg_aligns,\n",
    "    columns=[\n",
    "        'coverage_bp',\n",
    "        'contigs_count',\n",
    "        'delta',\n",
    "        'coverage_pct',\n",
    "        'num_gaps',\n",
    "        'min_gap',\n",
    "        'max_gap',\n",
    "        'avg_gap'\n",
    "    ],\n",
    "    index=pd.MultiIndex.from_tuples(\n",
    "        mrg_index,\n",
    "        names=['sample', 'platform', 'haplotype', 'roi']\n",
    "    )\n",
    ")\n",
    "\n",
    "hla_genes = [\n",
    "    'HLA-F',\n",
    "    'HLA-G',\n",
    "    'HLA-A',\n",
    "    'HLA-E',\n",
    "    'HLA-C',\n",
    "    'HLA-B',\n",
    "    'HLA-DRA',\n",
    "    'HLA-DRB5',\n",
    "    'HLA-DRB1',\n",
    "    'HLA-DQA1',\n",
    "    'HLA-DQB1',\n",
    "    'HLA-DQA2',\n",
    "    'HLA-DQB2',\n",
    "    'HLA-DOB',\n",
    "    'HLA-DMB',\n",
    "    'HLA-DMA',\n",
    "    'HLA-DOA',\n",
    "    'HLA-DPA1',\n",
    "    'HLA-DPB1',\n",
    "]\n",
    "\n",
    "drop_columns = [\n",
    "    'min_gap',\n",
    "    'max_gap',\n",
    "    'avg_gap',\n",
    "    'delta_percentile'\n",
    "]\n",
    "\n",
    "hla_table = []\n",
    "\n",
    "for roi in ['3q29', 'HLA'] + hla_genes:\n",
    "    print('################')\n",
    "    print('====', roi)\n",
    "    print('################')\n",
    "    subset = joined.xs(roi, level='roi', drop_level=False).copy()\n",
    "    subset.sort_values(['delta', 'coverage_pct', 'sample', 'platform', 'haplotype'],\n",
    "                       inplace=True,\n",
    "                       ascending=[True, False, True, True, True]\n",
    "                      )\n",
    "    subset['delta_percentile'] = (subset['delta'].rank(ascending=False, method='max', pct=True) * 100).round(2)\n",
    "    \n",
    "#    print('delta ', subset['delta'].describe())\n",
    "#    print('cov % ', subset['coverage_pct'].describe())\n",
    "#    print('ctg # ', subset['contigs_count'].describe())\n",
    "#    print('gaps # ', subset['num_gaps'].describe())\n",
    "    \n",
    "    if not roi.startswith('HLA-'):\n",
    "        continue\n",
    "    subset.drop(drop_columns, axis=1, inplace=True)\n",
    "    subset = subset.droplevel('roi', axis=0)\n",
    "    new_columns = [roi + '_' + c for c in subset.columns]\n",
    "    subset.columns = new_columns\n",
    "    hla_table.append(subset)\n",
    "    \n",
    "hla_table = pd.concat(hla_table, axis=1, ignore_index=False)\n",
    "hla_table.sort_index(axis=0, level=['platform', 'sample', 'haplotype'], ascending=True, inplace=True)\n",
    "hla_table.sort_index(axis=1, ascending=True, inplace=True)\n",
    "hla_table.fillna(0., inplace=True)\n",
    "#print(hla_table.head())\n",
    "\n",
    "# out_table = '/home/local/work/data/hgsvc/roi/roi_assembly_summary_HLAn.tsv'\n",
    "# hla_table.to_csv(\n",
    "#     out_table,\n",
    "#     sep='\\t',\n",
    "#     header=True,\n",
    "#     index=True\n",
    "# )\n",
    "\n",
    "cov_only = hla_table[[c for c in hla_table.columns if 'coverage_pct' in c]].copy()\n",
    "complete_hits = (cov_only > 99.999).sum(axis=0)\n",
    "partial_hits = ((0 < cov_only) & (cov_only < 99.999)).sum(axis=0)\n",
    "fails = (cov_only < 0.0001).sum(axis=0)\n",
    "\n",
    "condensed_table = pd.concat(\n",
    "    [complete_hits, partial_hits, fails],\n",
    "    axis=1,\n",
    ")\n",
    "condensed_table.columns = ['complete', 'partial', 'missing']\n",
    "condensed_table.index = [x.split('_')[0] for x in condensed_table.index]\n",
    "print(condensed_table)\n",
    "\n",
    "    #out_table = '/home/local/work/data/hgsvc/roi/roi_assembly_summary_{}.tsv'.format(roi)\n",
    "    #subset.to_csv(out_table, sep='\\t', header=True, index=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
