{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0.0, 0.0, 0.0)\n",
      "(1.0, 1.0, 1.0)\n",
      "(0.5333333333333333, 0.5333333333333333, 0.5333333333333333)\n",
      "  assembly_a assembly_b  haplotype_a  haplotype_b  intersect_bp  \\\n",
      "0        HAP        HAP            1            2       2609354   \n",
      "1        HAP        HAP            1            2       2191934   \n",
      "2        HAP        HAP            1            1       2244686   \n",
      "3        HAP        HAP            1            1       2645605   \n",
      "4        HAP        HAP            1            2         14658   \n",
      "\n",
      "   intersect_count   jaccard platform_a platform_b sample_a sample_b  \\\n",
      "0             2271  0.092653        CLR        CLR  HG00096  HG00171   \n",
      "1             1930  0.080305        CLR       HiFi  HG00096  HG00512   \n",
      "2             2059  0.081590        CLR       HiFi  HG00096  HG00512   \n",
      "3             2286  0.095327        CLR        CLR  HG00096  HG00171   \n",
      "4               34  0.000474        CLR        CLR  HG00096  HG00096   \n",
      "\n",
      "   union_bp  comparison marker    color family_a family_b super_a super_b  \n",
      "0  28162756           3      D  #2D6F91  HG00096  HG00171     EUR     EUR  \n",
      "1  27295238           3      s  #000000  HG00096    SH032     EUR     EAS  \n",
      "2  27511895           3      s  #000000  HG00096    SH032     EUR     EAS  \n",
      "3  27752844           3      D  #2D6F91  HG00096  HG00171     EUR     EUR  \n",
      "4  30897617           4      D  #2D6F91  HG00096  HG00096     EUR     EUR  \n",
      "     assembly_a assembly_b  haplotype_a  haplotype_b  intersect_bp  \\\n",
      "3316        HAP        HAP            2            1       2704987   \n",
      "3317        HAP        HAP            1            2         15043   \n",
      "3318        HAP        HAP            1            2       2594190   \n",
      "3319        HAP        HAP            2            2       2672559   \n",
      "3320        HAP        HAP            1            2          8659   \n",
      "\n",
      "      intersect_count   jaccard platform_a platform_b sample_a sample_b  \\\n",
      "3316             2269  0.093236        CLR       HiFi  NA20847  NA24385   \n",
      "3317               32  0.000454        CLR        CLR  NA20847  NA20847   \n",
      "3318             2184  0.088269        CLR       HiFi  NA20847  NA24385   \n",
      "3319             2124  0.090465        CLR       HiFi  NA20847  NA24385   \n",
      "3320               24  0.000283       HiFi       HiFi  NA24385  NA24385   \n",
      "\n",
      "      union_bp  comparison marker    color family_a family_b super_a super_b  \n",
      "3316  29012231           3      s  #000000  NA20847     3140     SAS     EUR  \n",
      "3317  33120757           4      D  #782B8A  NA20847  NA20847     SAS     SAS  \n",
      "3318  29389448           3      s  #000000  NA20847     3140     SAS     EUR  \n",
      "3319  29542543           3      s  #000000  NA20847     3140     SAS     EUR  \n",
      "3320  30556397           4      o  #2D6F91     3140     3140     EUR     EUR  \n"
     ]
    },
    {
     "ename": "RuntimeError",
     "evalue": "No active exception to reraise",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-3388bc362535>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     73\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     74\u001b[0m \u001b[0m_\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mload_jaccard_stats\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstats_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 75\u001b[0;31m \u001b[0;32mraise\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     76\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     77\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mextract_assembly_type\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mRuntimeError\u001b[0m: No active exception to reraise"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import collections as col\n",
    "import importlib\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as patches\n",
    "import matplotlib.lines as lines\n",
    "\n",
    "plot_aux_module = '/home/local/work/code/github/project-diploid-assembly/notebooks/aux_mods/plot_aux.py'\n",
    "plot_aux_spec = importlib.util.spec_from_file_location(\"plot_aux\", plot_aux_module)\n",
    "plot_aux = importlib.util.module_from_spec(plot_aux_spec)\n",
    "plot_aux_spec.loader.exec_module(plot_aux)\n",
    "\n",
    "samples = plot_aux.load_sample_table()\n",
    "hexcodes, rgbcodes, popmap = plot_aux.load_population_annotation()\n",
    "prop = plot_aux.get_plot_property\n",
    "pipeline_version = 'v12'\n",
    "\n",
    "stats_path = '/home/local/work/data/hgsvc/figSX_panels/hapreg_jaccard/hap_regs_jaccard.tsv'\n",
    "out_path = os.path.join(stats_path, pipeline_version, 'figSX_panel_hapreg_jaccard')\n",
    "cache_file = os.path.join(stats_path, 'cache_{}.h5'.format(pipeline_version))\n",
    "\n",
    "\n",
    "def load_jaccard_stats(file_path):\n",
    "    df = pd.read_csv(file_path, index_col=False, sep='\\t')\n",
    "    df['comparison'] = 0\n",
    "    df['marker'] = 's'\n",
    "    df['color'] = '#000000'\n",
    "    df['family_a'] = df['sample_a'].map(lambda x: samples[x]['family'])\n",
    "    df['family_b'] = df['sample_b'].map(lambda x: samples[x]['family'])\n",
    "    df['super_a'] = df['sample_a'].map(lambda x: samples[x]['super_population'])\n",
    "    df['super_b'] = df['sample_b'].map(lambda x: samples[x]['super_population'])\n",
    "    \n",
    "    # comparison HiFi vs CLR (same sample)\n",
    "    select_same_sample = df['sample_a'] == df['sample_b']\n",
    "    select_diff_tech = df['platform_a'] != df['platform_b']\n",
    "    df.loc[select_same_sample & select_diff_tech, 'comparison'] = 1\n",
    "    \n",
    "    # comparison parent vs child\n",
    "    select_diff_sample = df['sample_a'] != df['sample_b']\n",
    "    select_same_family = df['family_a'] == df['family_b']\n",
    "    df.loc[select_diff_sample & select_same_family, 'comparison'] = 2\n",
    "    \n",
    "    # comparison hap1 vs hap2\n",
    "    select_diff_hap = df['haplotype_a'] != df['haplotype_b']\n",
    "    df.loc[select_same_sample & select_diff_hap, 'comparison'] = 4\n",
    "    \n",
    "    # comparison \"any other\"\n",
    "    df.loc[df['comparison'] == 0, 'comparison'] = 3\n",
    "    \n",
    "    # marker CLR\n",
    "    select_both_clr = (df['platform_a'] == 'CLR') & (df['platform_b'] == 'CLR')\n",
    "    df.loc[select_both_clr, 'marker'] = prop('CLR_marker')\n",
    "    \n",
    "    # marker HiFi\n",
    "    select_both_hifi = (df['platform_a'] == 'HiFi') & (df['platform_b'] == 'HiFi')\n",
    "    df.loc[select_both_hifi, 'marker'] = prop('HiFi_marker')\n",
    "    \n",
    "    # assign color if both samples come from same super pop\n",
    "    df.loc[df['super_a'] == df['super_b'], 'color'] = \\\n",
    "        df.loc[df['super_a'] == df['super_b'], 'super_a'].map(lambda x: hexcodes[x])\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "def extract_assembly_type(file_name):\n",
    "    assmt = None\n",
    "    if '_nhr-' in file_name:\n",
    "        assmt = 'NHR', 0\n",
    "    elif 'h1-un' in file_name:\n",
    "        assmt = 'HAP', 1\n",
    "    elif 'h2-un' in file_name:\n",
    "        assmt = 'HAP', 2\n",
    "    else:\n",
    "        raise ValueError('ASSMT: {}'.format(file_name))\n",
    "    if assmt is None:\n",
    "        raise ValueError('ASSMT: {}'.format(file_name))\n",
    "    return assmt\n",
    "\n",
    "\n",
    "if not os.path.isfile(cache_file):\n",
    "    stat_files = plot_aux.load_plot_data_files(\n",
    "        stats_path,\n",
    "        '.txt',\n",
    "        pipeline_version)\n",
    "    plot_data = [load_assembly_stats(fp) for fp in stat_files]\n",
    "\n",
    "    sample_platform = [plot_aux.extract_sample_platform(os.path.basename(f)) for f in stat_files]\n",
    "    assembly_types = [extract_assembly_type(os.path.basename(f)) for f in stat_files]\n",
    "    row_index = []\n",
    "    for (sample, platform), (assmt, hap) in zip(sample_platform, assembly_types):\n",
    "        super_pop = samples[sample]['super_population']\n",
    "        pop = samples[sample]['population']\n",
    "        row_index.append((sample, super_pop, pop, platform, assmt, hap))\n",
    "\n",
    "    df = pd.DataFrame(\n",
    "        plot_data,\n",
    "        columns=sorted(plot_data[0].keys()),\n",
    "        index=row_index\n",
    "    )\n",
    "\n",
    "    df.index = pd.MultiIndex.from_tuples(\n",
    "        df.index.values,\n",
    "        names=['sample', 'super_pop', 'pop', 'platform', 'assembly', 'hap']\n",
    "    )\n",
    "    df.to_hdf(cache_file, key='cache', mode='w', format='fixed')\n",
    "    \n",
    "df = pd.read_hdf(cache_file, 'cache')\n",
    "df.sort_index(axis=0, inplace=True, level=['sample', 'super_pop', 'pop', 'platform', 'assembly', 'hap'])        \n",
    "\n",
    "    \n",
    "def plot_assembly_contiguity(sample_stats):\n",
    "    \n",
    "    nhr_assm = sample_stats.xs('NHR', level='assembly', drop_level=False)\n",
    "    \n",
    "    x_vals = col.defaultdict(list)\n",
    "    y_vals = col.defaultdict(list)\n",
    "    colors = col.defaultdict(list)\n",
    "    \n",
    "    for idx, row in nhr_assm.iterrows():\n",
    "        platform = idx[3]\n",
    "        population = idx[2]\n",
    "        \n",
    "        x_vals[platform].append(round(row['N50'] / 1e6, 2))\n",
    "        \n",
    "        hap1_index = (idx[0], idx[1], idx[2], idx[3], 'HAP', 1)\n",
    "        y_vals[(platform, 'HAP1')].append(round(sample_stats.loc[hap1_index, 'N50'] / 1e6, 2))\n",
    "        \n",
    "        hap2_index = (idx[0], idx[1], idx[2], idx[3], 'HAP', 2)\n",
    "        y_vals[(platform, 'HAP2')].append(round(sample_stats.loc[hap2_index, 'N50'] / 1e6, 2))\n",
    "        \n",
    "        colors[platform].append(rgbcodes[population])\n",
    "        \n",
    "    \n",
    "    fig, axis = plt.subplots(figsize=(8,8))\n",
    "    \n",
    "    for platform in plot_aux.get_sequencing_platforms():\n",
    "        for hap in ['HAP1', 'HAP2']:\n",
    "            nhr_n50 = x_vals[platform]\n",
    "            hap_n50 = y_vals[(platform, hap)]\n",
    "            marker_colors = colors[platform]\n",
    "\n",
    "            axis.scatter(\n",
    "                nhr_n50,\n",
    "                hap_n50,\n",
    "                s=prop('plot_marker_size'),\n",
    "                c=colors[platform],\n",
    "                marker=prop('{}_marker'.format(platform))\n",
    "            )\n",
    "\n",
    "    axis.set_xlabel(\n",
    "        'Squashed assembly contig N50 (Mbp)',\n",
    "        fontsize=prop('fontsize_legend')\n",
    "    )\n",
    "    axis.set_ylabel(\n",
    "        'Haploid assembly contig N50 (Mbp)',\n",
    "        fontsize=prop('fontsize_legend')\n",
    "    )\n",
    "    axis.spines['top'].set_visible(False)\n",
    "    axis.spines['right'].set_visible(False)\n",
    "    \n",
    "    axis.tick_params(\n",
    "        axis='both',\n",
    "        which='major',\n",
    "        labelsize=prop('fontsize_axis_ticks')\n",
    "    )\n",
    "    \n",
    "    \n",
    "    # build custom legend\n",
    "    custom_lines = [\n",
    "        lines.Line2D(\n",
    "            [0], [0],\n",
    "            color='black',\n",
    "            markersize=prop('legend_marker_size'),\n",
    "            marker=prop('CLR_marker'),\n",
    "            ls='None',\n",
    "            label='CLR'\n",
    "        ),\n",
    "        lines.Line2D(\n",
    "            [0], [0],\n",
    "            color='black',\n",
    "            markersize=prop('legend_marker_size'),\n",
    "            marker=prop('HiFi_marker'),\n",
    "            ls='None',\n",
    "            label='HiFi'\n",
    "        )   \n",
    "    ]\n",
    "    \n",
    "    axis.legend(\n",
    "        handles=custom_lines,\n",
    "        loc='lower right',\n",
    "        prop={'size': prop('fontsize_legend')}\n",
    "    )\n",
    "\n",
    "    if missing:\n",
    "        plot_aux.add_incomplete_stamp(axis, 0.5, 0.85)\n",
    "    \n",
    "    extra_artists = []\n",
    "    \n",
    "    return fig, extra_artists\n",
    "        \n",
    "fig, exart = plot_assembly_contiguity(df)\n",
    "    \n",
    "#fig.savefig(out_path + '.png', dpi=600, bbox_inches='tight', extra_artists=exart)\n",
    "#fig.savefig(out_path + '.svg', bbox_inches='tight', extra_artists=exart)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
