{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "RuntimeError",
     "evalue": "No active exception to reraise",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-14d2da742c1b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m    513\u001b[0m             mode='w')\n\u001b[1;32m    514\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 515\u001b[0;31m         \u001b[0;32mraise\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    516\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    517\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mrs\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mregion_select\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mRuntimeError\u001b[0m: No active exception to reraise"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import collections as col\n",
    "import importlib\n",
    "import re\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as patches\n",
    "import matplotlib.lines as lines\n",
    "\n",
    "plot_aux_module = '/home/local/work/code/github/project-diploid-assembly/notebooks/aux_mods/plot_aux.py'\n",
    "plot_aux_spec = importlib.util.spec_from_file_location(\"plot_aux\", plot_aux_module)\n",
    "plot_aux = importlib.util.module_from_spec(plot_aux_spec)\n",
    "plot_aux_spec.loader.exec_module(plot_aux)\n",
    "\n",
    "samples = plot_aux.load_sample_table()\n",
    "hexcodes, rgbcodes, popmap = plot_aux.load_population_annotation()\n",
    "prop = plot_aux.get_plot_property\n",
    "\n",
    "save_plots = False\n",
    "print_stats = True\n",
    "\n",
    "pipeline_version = 'v12'\n",
    "\n",
    "stats_path = '/home/local/work/data/hgsvc/figSX_panels/assm_completeness'\n",
    "out_path = os.path.join(stats_path, pipeline_version, 'figSX_panel_completeness')\n",
    "cache_file = os.path.join(stats_path, 'cache_{}.h5'.format(pipeline_version))\n",
    "\n",
    "\n",
    "def extract_region_type(region_name):\n",
    "    \n",
    "    if region_name.split('_')[0] in ['Gap', 'Variation', 'Unknown']:\n",
    "        return region_name.split('_')[0]\n",
    "    elif region_name[:3] in ['CEN', 'PAR', 'HET']:\n",
    "        return region_name[:3]\n",
    "    else:\n",
    "        raise ValueError('Cannot process: {}'.format(region_name))\n",
    "\n",
    "\n",
    "def load_coordinates():\n",
    "    local_paths = [\n",
    "        '/home/local/work/code/github/project-diploid-assembly/annotation/grch38/20200723_GRCh38_p13_regions.bed',\n",
    "        '/home/local/work/code/github/project-diploid-assembly/annotation/grch38/20200723_GRCh38_p13_unresolved-issues.bed',\n",
    "        '/home/local/work/code/github/project-diploid-assembly/annotation/grch38/GRCh38_cytobands.bed',\n",
    "        '/home/local/work/code/github/project-diploid-assembly/annotation/grch38/GRCh38_segdups.bed'\n",
    "    ]\n",
    "    coord_lut = dict()\n",
    "    for lp in local_paths:\n",
    "        label = os.path.basename(lp).split('.')[0]\n",
    "        df = pd.read_csv(lp, sep='\\t', usecols=[0, 1, 2, 3], skiprows=1,\n",
    "                         names=['chrom', 'start', 'end', 'name'])\n",
    "        coord_lut[label] = df\n",
    "    return coord_lut\n",
    "\n",
    "\n",
    "def load_sequence_profiles():\n",
    "    local_path = '/home/local/work/data/hgsvc/figSX_panels/region_nuc_profiles'\n",
    "    \n",
    "    names=[\n",
    "        'name',\n",
    "        'GC_frac',\n",
    "        'AT_frac',\n",
    "        'count_A',\n",
    "        'count_C',\n",
    "        'count_G',\n",
    "        'count_T',\n",
    "        'count_N',\n",
    "        'count_X',\n",
    "    ]\n",
    "    \n",
    "    access_lut = dict()\n",
    "    for fn in os.listdir(local_path):\n",
    "        fp = os.path.join(local_path, fn)\n",
    "        label = fn.split('.')[0].split('_', 3)[3]\n",
    "        df = pd.read_csv(fp, sep='\\t', usecols=[3, 4, 5, 6, 7, 8, 9, 10, 11], skiprows=1,\n",
    "                         names=names)\n",
    "        assert (df['count_X'] == 0).all(), 'Unrecognized nucleotide: {}'.format(fn)\n",
    "        df.drop('count_X', axis=1, inplace=True)\n",
    "        df['accessible_pct'] = ((df['GC_frac'] + df['AT_frac']) * 100).round(2)\n",
    "        access_lut[label] = df\n",
    "    return access_lut\n",
    "\n",
    "\n",
    "def add_profiles_to_cache(cache_path):\n",
    "    \n",
    "    with pd.HDFStore(cache_path, 'a') as hdf:\n",
    "        for key, value in load_sequence_profiles().items():\n",
    "            if 'issues' in key:\n",
    "                store_key = 'issues'\n",
    "            else:\n",
    "                store_key = key.rsplit('_', 1)[-1]\n",
    "            hdf.put(os.path.join('nucprof', store_key), value, format='fixed')\n",
    "    return\n",
    "\n",
    "\n",
    "def load_contig_coverages(path, accessibility):\n",
    "\n",
    "    contig_ovl_header = [\n",
    "        'chrom',\n",
    "        'start',\n",
    "        'end',\n",
    "        'name',\n",
    "        'ovl_chrom',\n",
    "        'ovl_start',\n",
    "        'ovl_end',\n",
    "        'coverage'\n",
    "    ]\n",
    "    \n",
    "    per_sample = []\n",
    "    for overlap_file in os.listdir(path):\n",
    "        file_path = os.path.join(path, overlap_file)\n",
    "        df = pd.read_csv(file_path, sep='\\t', names=contig_ovl_header,\n",
    "                         usecols=[\n",
    "                             'chrom',\n",
    "                             'start',\n",
    "                             'end',\n",
    "                             'name',\n",
    "                             'coverage'\n",
    "                         ])\n",
    "        total_ovl = df.groupby('name')['coverage'].sum()\n",
    "        df.drop('coverage', axis=1, inplace=True)\n",
    "        df.drop_duplicates('name', keep='first', inplace=True)\n",
    "        df = df.merge(total_ovl, on='name')\n",
    "        df['stat_type'] = 'count'\n",
    "        df['input_type'] = 'contig'\n",
    "        \n",
    "        sample, platform = plot_aux.extract_sample_platform(\n",
    "            overlap_file.split('_OVL_')[1],\n",
    "        )\n",
    "        \n",
    "        df['sample'] = sample\n",
    "        df['platform'] = platform\n",
    "        df['super_pop'] = samples[sample]['super_population']\n",
    "        df['population'] = samples[sample]['population']\n",
    "        df['assembly'] = 'HAP'\n",
    "        if 'h1-un' in overlap_file:\n",
    "            hap = 10\n",
    "        elif 'h2-un' in overlap_file:\n",
    "            hap = 20\n",
    "        else:\n",
    "            raise\n",
    "        df['hap'] = hap\n",
    "        \n",
    "        if 'cytobands' in overlap_file:\n",
    "            df['region_type'] = 'cytoband'\n",
    "        elif 'segdups' in overlap_file:\n",
    "            df['region_type'] = 'segdup'\n",
    "        else:\n",
    "            df['region_type'] = df['name'].map(extract_region_type)\n",
    "        \n",
    "        df2 = df.copy()\n",
    "        df2['stat_type'] = 'fraction'\n",
    "        df2['coverage'] = df2['coverage'] / (df2['end'] - df2['start'])\n",
    "        df2['coverage'] = (df2['coverage'] * 100).round(2)\n",
    "        \n",
    "        annotation_file = overlap_file.split('_OVL_')[0]\n",
    "        nuc_stats = accessibility[annotation_file]\n",
    "        df3 = df2.copy()\n",
    "        df3['stat_type'] = 'accessible'\n",
    "        df3 = df3.merge(nuc_stats, on='name', how='outer')\n",
    "        # coverage here is still fractional, copied from df2\n",
    "        df3['coverage'] = 100 - (df3['accessible_pct'] - df3['coverage']).abs().round(2)\n",
    "        df3.drop([\n",
    "            'GC_frac',\n",
    "            'AT_frac',\n",
    "            'count_A',\n",
    "            'count_C',\n",
    "            'count_G',\n",
    "            'count_T',\n",
    "            'count_N',\n",
    "            'accessible_pct'\n",
    "        ], axis=1, inplace=True)\n",
    "                \n",
    "        df = pd.concat([df, df2, df3], axis=0)\n",
    "        \n",
    "        index_infos = [\n",
    "            'chrom',\n",
    "            'region_type',\n",
    "            'stat_type',\n",
    "            'input_type',\n",
    "            'sample',\n",
    "            'population',\n",
    "            'super_pop',\n",
    "            'platform',\n",
    "            'assembly',\n",
    "            'hap'\n",
    "        ]\n",
    "        \n",
    "        # multiindex arrays\n",
    "        idx_arrays = [df[i] for i in index_infos]\n",
    "        idx_names = index_infos\n",
    "        \n",
    "        df.index = pd.MultiIndex.from_arrays(\n",
    "            idx_arrays,\n",
    "            names=idx_names\n",
    "        )\n",
    "        df.drop(index_infos, axis=1, inplace=True)\n",
    "        per_sample.append(df)\n",
    "\n",
    "    per_sample = pd.concat(per_sample, axis=0)\n",
    "\n",
    "    return per_sample\n",
    "     \n",
    "\n",
    "def load_read_coverages(path, coordinates, accessibility):\n",
    "    \n",
    "    read_cov_header = [\n",
    "        'name',\n",
    "        'length',\n",
    "        'coverage',\n",
    "        'coverage_cumsum',\n",
    "        'mean0',\n",
    "        'mean'\n",
    "    ]\n",
    "    \n",
    "    per_sample = []\n",
    "    for coverage_file in os.listdir(path):\n",
    "        file_path = os.path.join(path, coverage_file)\n",
    "        df = pd.read_csv(file_path, sep='\\t', names=read_cov_header)\n",
    "        \n",
    "        annotation_file, read_file = coverage_file.split('_AVG_')\n",
    "        \n",
    "        cov_count = df[['name', 'coverage']].copy()\n",
    "        cov_count['stat_type'] = 'count'\n",
    "        \n",
    "        cov_frac = df[['name', 'length', 'coverage']].copy()\n",
    "        cov_frac['stat_type'] = 'accessible'\n",
    "        cov_frac['coverage'] = cov_frac['coverage'] / cov_frac['length']\n",
    "        cov_frac['coverage'] = (cov_frac['coverage'] * 100).round(2)\n",
    "        cov_frac.drop('length', axis=1, inplace=True)\n",
    "        \n",
    "        nuc_stats = accessibility[annotation_file]\n",
    "        cov_access = cov_frac.copy()\n",
    "        cov_access['stat_type'] = 'accessible'\n",
    "        cov_access = cov_access.merge(nuc_stats, on='name', how='outer')\n",
    "        cov_access['coverage'] = 100 - (cov_access['accessible_pct'] - cov_access['coverage']).abs().round(2)\n",
    "        cov_access.drop([\n",
    "            'GC_frac',\n",
    "            'AT_frac',\n",
    "            'count_A',\n",
    "            'count_C',\n",
    "            'count_G',\n",
    "            'count_T',\n",
    "            'count_N',\n",
    "            'accessible_pct'\n",
    "        ], axis=1, inplace=True)\n",
    "        \n",
    "        cov_mean0 = df[['name', 'mean0']].copy()\n",
    "        cov_mean0['stat_type'] = 'average'\n",
    "        cov_mean0.columns = ['name', 'coverage', 'stat_type']\n",
    "        \n",
    "        cov_mean = df[['name', 'mean']].copy()\n",
    "        cov_mean['stat_type'] = 'nz_avg'\n",
    "        cov_mean.columns = ['name', 'coverage', 'stat_type']\n",
    "        \n",
    "        df = pd.concat([cov_count, cov_frac, cov_access, cov_mean0, cov_mean], axis=0)\n",
    "        \n",
    "        coords = coordinates[annotation_file]\n",
    "                \n",
    "        sample, platform = plot_aux.extract_sample_platform(read_file)\n",
    "        \n",
    "        df['input_type'] = 'read'\n",
    "        df['sample'] = sample\n",
    "        df['platform'] = platform\n",
    "        df['super_pop'] = samples[sample]['super_population']\n",
    "        df['population'] = samples[sample]['population']\n",
    "        df['assembly'] = 'HAP'\n",
    "        if 'h1.tab' in coverage_file:\n",
    "            hap = 1\n",
    "        elif 'h2.tab' in coverage_file:\n",
    "            hap = 2\n",
    "        elif '.un.tab' in coverage_file:\n",
    "            hap = 0\n",
    "        elif 'h1-un.tab' in coverage_file:\n",
    "            hap = 10\n",
    "        elif 'h2-un.tab' in coverage_file:\n",
    "            hap = 20\n",
    "        else:\n",
    "            raise\n",
    "\n",
    "        df['hap'] = hap\n",
    "        if 'cytobands' in coverage_file:\n",
    "            df['region_type'] = 'cytoband'\n",
    "        elif 'segdups' in coverage_file:\n",
    "            df['region_type'] = 'segdup'\n",
    "        else:\n",
    "            df['region_type'] = df['name'].map(extract_region_type)\n",
    "        \n",
    "        df = df.merge(coords, on='name', how='inner')\n",
    "        \n",
    "        index_infos = [\n",
    "            'chrom',\n",
    "            'region_type',\n",
    "            'stat_type',\n",
    "            'input_type',\n",
    "            'sample',\n",
    "            'population',\n",
    "            'super_pop',\n",
    "            'platform',\n",
    "            'assembly',\n",
    "            'hap'\n",
    "        ]\n",
    "        \n",
    "        # multiindex arrays\n",
    "        idx_arrays = [df[i] for i in index_infos]\n",
    "        idx_names = index_infos\n",
    "        \n",
    "        df.index = pd.MultiIndex.from_arrays(\n",
    "            idx_arrays,\n",
    "            names=idx_names\n",
    "        )\n",
    "        df.drop(index_infos, axis=1, inplace=True)\n",
    "        per_sample.append(df)\n",
    "\n",
    "    per_sample = pd.concat(per_sample, axis=0)\n",
    "\n",
    "    return per_sample\n",
    "\n",
    "\n",
    "if not os.path.isfile(cache_file):\n",
    "    \n",
    "    contig_stats = load_contig_coverages(\n",
    "        os.path.join(stats_path, 'contig_coverage'),\n",
    "        load_sequence_profiles()\n",
    "    )\n",
    "    read_stats = load_read_coverages(\n",
    "        os.path.join(stats_path, 'hap_read_coverage'),\n",
    "        load_coordinates(),\n",
    "        load_sequence_profiles()\n",
    "    )\n",
    "    \n",
    "    df = pd.concat([contig_stats, read_stats], axis=0)\n",
    "    \n",
    "    df.to_hdf(cache_file, key='cache', mode='w', format='fixed')\n",
    "    add_profiles_to_cache(cache_file)\n",
    "    \n",
    "df = pd.read_hdf(cache_file, 'cache')\n",
    "     \n",
    "main_chroms = ['chr' + str(i) for i in range(1, 23)]\n",
    "main_chroms.append('chrX')\n",
    "df = df.loc[df.index.isin(main_chroms, level='chrom'), :]\n",
    "\n",
    "missing=False\n",
    "\n",
    "def plot_contig_coverage(sample_stats, stat_type):\n",
    "    \n",
    "    clr_assm = sample_stats.xs(\n",
    "        ['CLR', 'contig', 'cytoband', stat_type],\n",
    "        level=['platform', 'input_type', 'region_type', 'stat_type'],\n",
    "        drop_level=True\n",
    "    ).copy()\n",
    "    clr_cov = pd.DataFrame(clr_assm.groupby(['chrom', 'start', 'name'])['coverage'].mean())\n",
    "    clr_cov['sort_order'] = clr_cov.index.get_level_values('chrom').map(lambda x: 23 if x == 'chrX' else int(x.strip('chr')))\n",
    "    clr_cov.sort_values(['sort_order', 'start'], ascending=True, inplace=True)\n",
    "    #clr_std = clr_assm.groupby(['chrom', 'start', 'name'])['coverage'].std()\n",
    "                \n",
    "    hifi_assm = sample_stats.xs(\n",
    "        ['HiFi', 'contig', 'cytoband', stat_type],\n",
    "        level=['platform', 'input_type', 'region_type', 'stat_type'],\n",
    "        drop_level=True\n",
    "    )\n",
    "    hifi_cov = pd.DataFrame(hifi_assm.groupby(['chrom', 'start', 'name'])['coverage'].mean())\n",
    "    hifi_cov['sort_order'] = hifi_cov.index.get_level_values('chrom').map(lambda x: 23 if x == 'chrX' else int(x.strip('chr')))\n",
    "    hifi_cov.sort_values(['sort_order', 'start'], ascending=True, inplace=True)\n",
    "    \n",
    "    #hifi_std = hifi_assm.groupby(['chrom', 'start', 'name'])['coverage'].std()\n",
    "\n",
    "    \n",
    "    fig, axis = plt.subplots(figsize=(16,8))\n",
    "    \n",
    "    x_vals_clr = np.array(range(0, clr_cov.shape[0]*2, 2), dtype=np.float16)\n",
    "    x_vals_hifi = x_vals_clr + 1\n",
    "\n",
    "    axis.errorbar(\n",
    "        x_vals_clr,\n",
    "        clr_cov['coverage'].values,\n",
    "        yerr=None, #clr_std.values,\n",
    "        marker=prop('CLR_marker'),\n",
    "        color=plot_aux.get_platform_color('CLR'),\n",
    "        linestyle='',\n",
    "        markersize=4\n",
    "    )\n",
    "    \n",
    "    axis.errorbar(\n",
    "        x_vals_hifi,\n",
    "        hifi_cov['coverage'].values,\n",
    "        yerr=None,#hifi_std.values,\n",
    "        marker=prop('HiFi_marker'),\n",
    "        color=plot_aux.get_platform_color('HiFi'),\n",
    "        linestyle='',\n",
    "        markersize=4,\n",
    "        alpha=0.75\n",
    "    )\n",
    "    \n",
    "    x_ticks = []\n",
    "    x_ticklabels = []\n",
    "    vlines_boundary = []\n",
    "    vlines_cen = []\n",
    "    cytogenic_bands = clr_cov.index.get_level_values('name')\n",
    "    last_chrom = cytogenic_bands[0].split('p')[0]\n",
    "    last_arm = 'p'\n",
    "    extract = re.compile('(?P<chrom>chr[0-9X]+)(?P<arm>(p|q))(?P<segment>[0-9\\.]+)')\n",
    "    for x_val, band_id in zip(x_vals_hifi, cytogenic_bands):\n",
    "        mobj = extract.match(band_id)\n",
    "        if mobj is None:\n",
    "            raise ValueError(band_id)\n",
    "        chrom = mobj.group('chrom')\n",
    "        arm = mobj.group('arm')\n",
    "        if chrom != last_chrom:\n",
    "            vlines_boundary.append(x_val + 0.5)\n",
    "            last_chrom = chrom\n",
    "            last_arm = 'p'\n",
    "            continue\n",
    "        elif arm != last_arm:\n",
    "            x_ticks.append(x_val + 0.5)\n",
    "            x_ticklabels.append(chrom)\n",
    "            vlines_cen.append(x_val + 0.5)\n",
    "            last_arm = arm\n",
    "        else:\n",
    "            pass\n",
    "    \n",
    "    axis.vlines(vlines_boundary, 90, 105, color='black', linestyle='solid', zorder=0)\n",
    "    axis.vlines(vlines_cen, 0, 100, color='dimgrey', linestyle='dotted', zorder=0)\n",
    "    \n",
    "    axis.set_xticks(x_ticks)\n",
    "    axis.set_xticklabels([x.strip('chr') for x in x_ticklabels], rotation=90)\n",
    "\n",
    "    axis.set_xlabel(\n",
    "        'Genomic location / cytogenetic bands',\n",
    "        fontsize=prop('fontsize_legend')\n",
    "    )\n",
    "    axis.set_ylabel(\n",
    "        'Haploid assembly contig coverage\\n(average, % bp)',\n",
    "        fontsize=prop('fontsize_legend')\n",
    "    )\n",
    "    axis.spines['top'].set_visible(False)\n",
    "    axis.spines['right'].set_visible(False)\n",
    "    \n",
    "    axis.tick_params(\n",
    "        axis='both',\n",
    "        which='major',\n",
    "        labelsize=prop('fontsize_axis_ticks')\n",
    "    )\n",
    "        \n",
    "    # build custom legend\n",
    "    custom_lines = [\n",
    "        lines.Line2D(\n",
    "            [0], [0],\n",
    "            color=plot_aux.get_platform_color('CLR'),\n",
    "            markersize=prop('legend_marker_size'),\n",
    "            marker=prop('CLR_marker'),\n",
    "            ls='None',\n",
    "            label='CLR',\n",
    "        ),\n",
    "        lines.Line2D(\n",
    "            [0], [0],\n",
    "            color=plot_aux.get_platform_color('HiFi'),\n",
    "            markersize=prop('legend_marker_size'),\n",
    "            marker=prop('HiFi_marker'),\n",
    "            ls='None',\n",
    "            label='HiFi',\n",
    "        )   \n",
    "    ]\n",
    "    \n",
    "    lgd = axis.legend(\n",
    "        handles=custom_lines,\n",
    "        prop={'size': prop('fontsize_legend')},\n",
    "        bbox_to_anchor=(0.95, 0.75)\n",
    "    )\n",
    "\n",
    "    if missing:\n",
    "        plot_aux.add_incomplete_stamp(axis, 0.5, 0.85)\n",
    "    \n",
    "    extra_artists = [lgd]\n",
    "    \n",
    "    return fig, extra_artists\n",
    "\n",
    "# fraction: % covered bp, not normalizing for N\n",
    "# accessible: % covered bp, normalizing for N\n",
    "\n",
    "for stat_type in ['fraction', 'accessible']:\n",
    "    print('### STATISTIC: ', stat_type)\n",
    "    if save_plots:\n",
    "        fig, exart = plot_contig_coverage(df, stat_type)\n",
    "\n",
    "        fig.savefig(\n",
    "            out_path + '_' + stat_type + '_lowres.png',\n",
    "            dpi=prop('dpi_low_res'),\n",
    "            bbox_inches='tight',\n",
    "            extra_artists=exart\n",
    "        )\n",
    "        fig.savefig(\n",
    "            out_path + '_' + stat_type + '.png',\n",
    "            dpi=prop('dpi_high_res'),\n",
    "            bbox_inches='tight',\n",
    "            extra_artists=exart\n",
    "        )\n",
    "        fig.savefig(out_path + '_' + stat_type + '.svg', bbox_inches='tight', extra_artists=exart)\n",
    "\n",
    "    if print_stats:\n",
    "        region_select = [\n",
    "            'CEN',\n",
    "            'Gap',\n",
    "            'segdup'\n",
    "        ]\n",
    "        \n",
    "        for rs in region_select:\n",
    "            if rs == 'Gap':\n",
    "                subsets = [\n",
    "                    df.xs(\n",
    "                        ['contig', rs, stat_type],\n",
    "                        level=['input_type', 'region_type', 'stat_type']\n",
    "                    ),\n",
    "                    df.xs(\n",
    "                        ['contig', 'Unknown', stat_type],\n",
    "                        level=['input_type', 'region_type', 'stat_type']\n",
    "                    )\n",
    "                ]\n",
    "                subsets = pd.concat(subsets, ignore_index=False, axis=0)\n",
    "                print('---> Region: {}'.format(rs))\n",
    "                print('=== ALL stats summary')\n",
    "                print(subsets['coverage'].describe())\n",
    "\n",
    "                clr_stats = subsets.xs(\n",
    "                    ['CLR'],\n",
    "                    level=['platform']\n",
    "                )\n",
    "                print('=== CLR stats')\n",
    "                print(clr_stats['coverage'].describe())\n",
    "\n",
    "                hifi_stats = subsets.xs(\n",
    "                    ['HiFi'],\n",
    "                    level=['platform']\n",
    "                )\n",
    "                print('=== HiFi stats')\n",
    "                print(hifi_stats['coverage'].describe())\n",
    "\n",
    "            else:\n",
    "                print('---> Region: {}'.format(rs))\n",
    "                print('=== ALL stats summary')\n",
    "                all_stats = df.xs(\n",
    "                    ['contig', rs, stat_type],\n",
    "                    level=['input_type', 'region_type', 'stat_type']\n",
    "                )\n",
    "                print(all_stats['coverage'].describe())\n",
    "\n",
    "                clr_stats = df.xs(\n",
    "                    ['CLR', 'contig', rs, stat_type],\n",
    "                    level=['platform', 'input_type', 'region_type', 'stat_type']\n",
    "                )\n",
    "                print('=== CLR stats')\n",
    "                print(clr_stats['coverage'].describe())\n",
    "\n",
    "                hifi_stats = df.xs(\n",
    "                    ['HiFi', 'contig', rs, stat_type],\n",
    "                    level=['platform', 'input_type', 'region_type', 'stat_type']\n",
    "                )\n",
    "                print('=== HiFi stats')\n",
    "                print(hifi_stats['coverage'].describe())\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
