{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== STATS SUMMARY ===\n",
      "= ALL\n",
      "count    966.000000\n",
      "mean      98.433230\n",
      "std        4.568985\n",
      "min       36.250000\n",
      "25%       98.710000\n",
      "50%       99.410000\n",
      "75%       99.760000\n",
      "max      100.550000\n",
      "Name: top1_cluster_align_pct, dtype: float64\n",
      "---\n",
      "= CLR\n",
      "count    690.000000\n",
      "mean      98.972464\n",
      "std        1.982397\n",
      "min       75.690000\n",
      "25%       98.982500\n",
      "50%       99.490000\n",
      "75%       99.830000\n",
      "max      100.550000\n",
      "Name: top1_cluster_align_pct, dtype: float64\n",
      "---\n",
      "= HiFi\n",
      "count    276.000000\n",
      "mean      97.085145\n",
      "std        7.800947\n",
      "min       36.250000\n",
      "25%       97.757500\n",
      "50%       99.095000\n",
      "75%       99.562500\n",
      "max       99.960000\n",
      "Name: top1_cluster_align_pct, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import collections as col\n",
    "import importlib\n",
    "import re\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as patches\n",
    "import matplotlib.lines as lines\n",
    "\n",
    "plot_aux_module = '/home/local/work/code/github/project-diploid-assembly/notebooks/aux_mods/plot_aux.py'\n",
    "plot_aux_spec = importlib.util.spec_from_file_location(\"plot_aux\", plot_aux_module)\n",
    "plot_aux = importlib.util.module_from_spec(plot_aux_spec)\n",
    "plot_aux_spec.loader.exec_module(plot_aux)\n",
    "\n",
    "samples = plot_aux.load_sample_table()\n",
    "hexcodes, rgbcodes, popmap = plot_aux.load_population_annotation()\n",
    "prop = plot_aux.get_plot_property\n",
    "\n",
    "save_plots = False\n",
    "print_stats = True\n",
    "\n",
    "pipeline_version = 'v12'\n",
    "\n",
    "stats_path = '/home/local/work/data/hgsvc/figSX_panels/contig_stats'\n",
    "out_path = os.path.join(stats_path, pipeline_version, 'figSX_contig_aln_stats')\n",
    "cache_file = os.path.join(stats_path, 'cache_{}.h5'.format(pipeline_version))\n",
    "\n",
    "\n",
    "def split_cluster_alignment_info(record, index):\n",
    "    names = [\n",
    "        'top{}_cluster_name',\n",
    "        'top{}_cluster_length',\n",
    "        'top{}_cluster_align_pct',\n",
    "        'top{}_ref_align_pct'\n",
    "    ]\n",
    "    keys = [n.format(index) for n in names]\n",
    "    cast_types = [str, int, float, float]\n",
    "    if not isinstance(record, str):\n",
    "        assert pd.isna(record), 'Error: {}'.format(record)\n",
    "        values = 'no_align 0 0 0'.split()\n",
    "    else:\n",
    "        values = record.split('|')\n",
    "    d = dict((k, t(v)) for k, v, t in zip(keys, values, cast_types))\n",
    "    return d\n",
    "\n",
    "\n",
    "def extract_assembly_haplotype(filename):\n",
    "    \n",
    "    if 'h1-un' in filename:\n",
    "        assm, hap = 'HAP', 10\n",
    "    elif 'h2-un' in filename:\n",
    "        assm, hap = 'HAP', 20\n",
    "    elif 'h1' in filename:\n",
    "        assm, hap = 'HAP', 1\n",
    "    elif 'h2' in filename:\n",
    "        assm, hap = 'HAP', 2\n",
    "    elif 'nhr' in filename:\n",
    "        assm, hap = 'NHR', 0\n",
    "    elif 'scV{}'.format(pipeline_version.strip('v')) in filename:\n",
    "        assm, hap = 'NHR', 0\n",
    "    else:\n",
    "        raise ValueError('cannot extract assembly {}'.format(filename))\n",
    "    return assm, hap\n",
    "\n",
    "\n",
    "def load_contig_align_stats(path):\n",
    "   \n",
    "    index_infos = [\n",
    "        'chrom',\n",
    "        'sample',\n",
    "        'population',\n",
    "        'super_pop',\n",
    "        'platform',\n",
    "        'assembly',\n",
    "        'hap'\n",
    "    ]\n",
    "\n",
    "    per_sample = []\n",
    "    for stats_file in os.listdir(path):\n",
    "        if not stats_file.endswith('mapq60.stats'):\n",
    "            continue\n",
    "        sample, platform = plot_aux.extract_sample_platform(\n",
    "            stats_file,\n",
    "            mapped_readset=True,\n",
    "            long_read_pos=1\n",
    "        )\n",
    "        assm, hap = extract_assembly_haplotype(stats_file)\n",
    "        file_path = os.path.join(path, stats_file)\n",
    "        \n",
    "        df = pd.read_csv(file_path, sep='\\t')\n",
    "        df.columns = [c.strip('#') for c in df.columns]\n",
    "        \n",
    "        top_alignments = []\n",
    "        for i in range(1, 4):\n",
    "            top_aln = df['top{}_alignment'.format(i)].apply(split_cluster_alignment_info, args=(i,))\n",
    "            if len(top_alignments) > 0:\n",
    "                [d.update(e) for d, e in zip(top_alignments, top_aln)]\n",
    "            else:\n",
    "                top_alignments = top_aln\n",
    "            \n",
    "        top_alignments = pd.DataFrame.from_records(top_alignments)\n",
    "        df = df[['ref_seq', 'ref_length']]\n",
    "        df = pd.concat([df, top_alignments], axis=1)\n",
    "        df['sample'] = sample\n",
    "        df['platform'] = platform\n",
    "        df['super_pop'] = samples[sample]['super_population']\n",
    "        df['population'] = samples[sample]['population']\n",
    "        df['assembly'] = assm\n",
    "        df['hap'] = hap\n",
    "        df['chrom'] = df['ref_seq']\n",
    "        df.drop('ref_seq', axis=1, inplace=True)\n",
    "                \n",
    "        # multiindex arrays\n",
    "        idx_arrays = [df[i] for i in index_infos]\n",
    "        idx_names = index_infos\n",
    "        \n",
    "        df.index = pd.MultiIndex.from_arrays(\n",
    "            idx_arrays,\n",
    "            names=idx_names\n",
    "        )\n",
    "        df.drop(index_infos, axis=1, inplace=True)\n",
    "        per_sample.append(df)\n",
    "\n",
    "    per_sample = pd.concat(per_sample, axis=0)\n",
    "    \n",
    "    return per_sample\n",
    "\n",
    "\n",
    "if not os.path.isfile(cache_file):\n",
    "    \n",
    "    df = load_contig_align_stats(os.path.join(stats_path, 'clustered'))       \n",
    "    df.to_hdf(cache_file, key='cache', mode='w', format='fixed')\n",
    "    \n",
    "df = pd.read_hdf(cache_file, 'cache')\n",
    "     \n",
    "main_chroms = ['chr' + str(i) for i in range(1, 23)]\n",
    "main_chroms.append('chrX')\n",
    "\n",
    "df = df.loc[df.index.isin(main_chroms, level='chrom'), :]\n",
    "\n",
    "def plot_contig_coverage(sample_stats):\n",
    "    \n",
    "    clr_assm = sample_stats.xs(\n",
    "        ['CLR', 'contig', 'cytoband', 'fraction'],\n",
    "        level=['platform', 'input_type', 'region_type', 'stat_type'],\n",
    "        drop_level=True\n",
    "    ).copy()\n",
    "    clr_cov = pd.DataFrame(clr_assm.groupby(['chrom', 'start', 'name'])['coverage'].mean())\n",
    "    clr_cov['sort_order'] = clr_cov.index.get_level_values('chrom').map(lambda x: 23 if x == 'chrX' else int(x.strip('chr')))\n",
    "    clr_cov.sort_values(['sort_order', 'start'], ascending=True, inplace=True)\n",
    "    #clr_std = clr_assm.groupby(['chrom', 'start', 'name'])['coverage'].std()\n",
    "                \n",
    "    hifi_assm = sample_stats.xs(\n",
    "        ['HiFi', 'contig', 'cytoband', 'fraction'],\n",
    "        level=['platform', 'input_type', 'region_type', 'stat_type'],\n",
    "        drop_level=True\n",
    "    )\n",
    "    hifi_cov = pd.DataFrame(hifi_assm.groupby(['chrom', 'start', 'name'])['coverage'].mean())\n",
    "    hifi_cov['sort_order'] = hifi_cov.index.get_level_values('chrom').map(lambda x: 23 if x == 'chrX' else int(x.strip('chr')))\n",
    "    hifi_cov.sort_values(['sort_order', 'start'], ascending=True, inplace=True)\n",
    "    \n",
    "    #hifi_std = hifi_assm.groupby(['chrom', 'start', 'name'])['coverage'].std()\n",
    "\n",
    "    \n",
    "    fig, axis = plt.subplots(figsize=(16,8))\n",
    "    \n",
    "    x_vals_clr = np.array(range(0, clr_cov.shape[0]*2, 2), dtype=np.float16)\n",
    "    x_vals_hifi = x_vals_clr + 1\n",
    "\n",
    "    axis.errorbar(\n",
    "        x_vals_clr,\n",
    "        clr_cov['coverage'].values,\n",
    "        yerr=None, #clr_std.values,\n",
    "        marker=prop('CLR_marker'),\n",
    "        color='black',\n",
    "        linestyle='',\n",
    "        markersize=4\n",
    "    )\n",
    "    \n",
    "    axis.errorbar(\n",
    "        x_vals_hifi,\n",
    "        hifi_cov['coverage'].values,\n",
    "        yerr=None,#hifi_std.values,\n",
    "        marker=prop('HiFi_marker'),\n",
    "        color='red',\n",
    "        linestyle='',\n",
    "        markersize=4,\n",
    "        alpha=0.75\n",
    "    )\n",
    "    \n",
    "    x_ticks = []\n",
    "    x_ticklabels = []\n",
    "    vlines_boundary = []\n",
    "    vlines_cen = []\n",
    "    cytogenic_bands = clr_cov.index.get_level_values('name')\n",
    "    last_chrom = cytogenic_bands[0].split('p')[0]\n",
    "    last_arm = 'p'\n",
    "    extract = re.compile('(?P<chrom>chr[0-9X]+)(?P<arm>(p|q))(?P<segment>[0-9\\.]+)')\n",
    "    for x_val, band_id in zip(x_vals_hifi, cytogenic_bands):\n",
    "        mobj = extract.match(band_id)\n",
    "        if mobj is None:\n",
    "            raise ValueError(band_id)\n",
    "        chrom = mobj.group('chrom')\n",
    "        arm = mobj.group('arm')\n",
    "        if chrom != last_chrom:\n",
    "            vlines_boundary.append(x_val + 0.5)\n",
    "            last_chrom = chrom\n",
    "            last_arm = 'p'\n",
    "            continue\n",
    "        elif arm != last_arm:\n",
    "            x_ticks.append(x_val + 0.5)\n",
    "            x_ticklabels.append(chrom)\n",
    "            vlines_cen.append(x_val + 0.5)\n",
    "            last_arm = arm\n",
    "        else:\n",
    "            pass\n",
    "    \n",
    "    axis.vlines(vlines_boundary, 90, 105, color='black', linestyle='solid', zorder=0)\n",
    "    axis.vlines(vlines_cen, 0, 100, color='dimgrey', linestyle='dotted', zorder=0)\n",
    "    \n",
    "    axis.set_xticks(x_ticks)\n",
    "    axis.set_xticklabels(x_ticklabels, rotation=90)\n",
    "\n",
    "    axis.set_xlabel(\n",
    "        'Genomic location',\n",
    "        fontsize=prop('fontsize_legend')\n",
    "    )\n",
    "    axis.set_ylabel(\n",
    "        'Phased assembly contig coverage\\n(average % bp)',\n",
    "        fontsize=prop('fontsize_legend')\n",
    "    )\n",
    "    axis.spines['top'].set_visible(False)\n",
    "    axis.spines['right'].set_visible(False)\n",
    "    \n",
    "    axis.tick_params(\n",
    "        axis='both',\n",
    "        which='major',\n",
    "        labelsize=prop('fontsize_axis_ticks')\n",
    "    )\n",
    "        \n",
    "    # build custom legend\n",
    "    custom_lines = [\n",
    "        lines.Line2D(\n",
    "            [0], [0],\n",
    "            color='black',\n",
    "            markersize=prop('legend_marker_size'),\n",
    "            marker=prop('CLR_marker'),\n",
    "            ls='None',\n",
    "            label='CLR',\n",
    "        ),\n",
    "        lines.Line2D(\n",
    "            [0], [0],\n",
    "            color='red',\n",
    "            markersize=prop('legend_marker_size'),\n",
    "            marker=prop('HiFi_marker'),\n",
    "            ls='None',\n",
    "            label='HiFi',\n",
    "        )   \n",
    "    ]\n",
    "    \n",
    "    lgd = axis.legend(\n",
    "        handles=custom_lines,\n",
    "        prop={'size': prop('fontsize_legend')},\n",
    "        bbox_to_anchor=(0.95, 0.75)\n",
    "    )\n",
    "\n",
    "    if missing:\n",
    "        plot_aux.add_incomplete_stamp(axis, 0.5, 0.85)\n",
    "    \n",
    "    extra_artists = [lgd]\n",
    "    \n",
    "    return fig, extra_artists\n",
    "\n",
    "\n",
    "if save_plots:\n",
    "    fig, exart = plot_contig_coverage(df)\n",
    "    fig.savefig(out_path + '.png', dpi=600, bbox_inches='tight', extra_artists=exart)\n",
    "    fig.savefig(out_path + '.svg', bbox_inches='tight', extra_artists=exart)\n",
    "    \n",
    "if print_stats:\n",
    "    print('=== STATS SUMMARY ===')\n",
    "    print('= ALL')\n",
    "    print(df['top1_cluster_align_pct'].describe())\n",
    "    print('---')\n",
    "    subset_clr = df.xs('CLR', level='platform')\n",
    "    print('= CLR')\n",
    "    print(subset_clr['top1_cluster_align_pct'].describe())\n",
    "    print('---')\n",
    "    subset_hifi = df.xs('HiFi', level='platform')\n",
    "    print('= HiFi')\n",
    "    print(subset_hifi['top1_cluster_align_pct'].describe())\n",
    "    \n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
